# Minor allele frequency estimation per every population and morph group
```{r}
#!/bin/bash -l

#SBATCH -A naiss2023-5-222
#SBATCH -p core
#SBATCH -n 8
#SBATCH -t 04-00:00:00
#SBATCH --array=1-12
#SBATCH -J maf
#SBATCH -e maf_%A_%a.err
#SBATCH -o maf_%A_%a.out
#SBATCH --mail-type=all
#SBATCH --mail-user=khrystyna.kurta@imbim.uu.se

#Load modules
module load bioinfo-tools
module load ANGSD/0.933


#STEP 1: Define paths to Refference genome
# Update this with the name of the ref fasta file
#STEP 1: Define paths to Refference genome
REFGENOME=/proj/snic2020-2-19/private/arctic_charr/assemblies/fSalAlp1.1/Data_Package_fSalAlp1_assembly_20231024/fSalAlp1.1.hap1.cur.20231016.fasta

REF_INDEXED=/proj/snic2020-2-19/private/arctic_charr/assemblies/fSalAlp1.1/Data_Package_fSalAlp1_assembly_20231024/fSalAlp1.1.hap1.cur.20231016.fasta.fai


#STEP 2: Set up PATH to bam lists 
BAM_LIST_PATH=/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/BAM_LISTS/MAF_list

#STEP 3: Specify PATH to sites 
SITES=/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/SITES/sites_sorted.txt

#STEP 4: Specify which bam to use
BAM_LIST=$(ls $BAM_LIST_PATH/maf_*.list | sed -n ${SLURM_ARRAY_TASK_ID}p)

BAM_TARGET=${BAM_LIST/.list/}
OUTPUT=$(basename $BAM_TARGET)

#STEP 5: Specify the number of individuals
x=`cat $BAM_LIST | wc -l`

#STEP 6: Specify out directory
OUT_DIR=/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/MAF

#STEP 7: Go to the bam files directory
BASEDIR=/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/Run_1_2
cd $BASEDIR

#STEP 8: Run the code:
echo "Estimating maf for $x number of individuals."

#Index sites 
#angsd sites index $SITES

angsd -out $OUT_DIR/$OUTPUT \
-ref $REFGENOME -fai $REF_INDEXED \
-doMajorMinor 4 -doMaf 1 -bam $BAM_LIST -nInd $x \
-uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -minMapQ 30 -minQ 20 \
-gl 2 \
-sites $SITES \
-nThreads 8
```

