# Allele frequency of diagnostic SNPs outside the putative inversions selected based on the benthic/pelagic contrast of Arctic charr morphs from Lake Thingvallavatn.

## Step 1: Load MAF file 
```{r }
dir = '~/Desktop/Comp_UU/REF_SalAlp_UK/MAF'

#Start
library(tidyverse)
library(dplyr)
library(ggplot2)
library(ggpubr)
options(scipen = 999)
library(scales)
library(ComplexHeatmap)
library(purrr)
library(psych)


#This to prepare Lake_Morph for the heatmap


#Read MAF for all pops
all_maf_filter <- read.table(file.path(dir, "Maf_all_pops.txt"), header = T)
all_maf_filter$SNP_ID <- paste(all_maf_filter$chromo,all_maf_filter$position, sep = "_")

#Add morph info
all_maf_filter <- 
  all_maf_filter %>%  
  mutate(Lake = case_when(substr(Lake_Morph, 1, 3) == "Myv" ~ "Mývatn",
                          substr(Lake_Morph, 1, 3) == "Thi" ~ "Thingvallavatn", 
                          substr(Lake_Morph, 1, 3) == "Sir" ~ "Sirdalsvatnet",
                          substr(Lake_Morph, 1, 3) == "Van"~ "Vangsvatnet" )
  )

```


## Step 2: Select outlier SNPs based on B vs P contrast
```{r }
Thin_LB_SB_vs_Pi_PL <- read.table('~/Desktop/Comp_UU/REF_SalAlp_UK/GWAS/Data/Thi_DB_LB_vs_PL_Pi.pvalue', header = T)
Thin_LB_SB_vs_Pi_PL$SNP_ID <- paste(Thin_LB_SB_vs_Pi_PL$Chromosome,Thin_LB_SB_vs_Pi_PL$Position, sep = "_")

bonf_1 = 1e-8/nrow(Thin_LB_SB_vs_Pi_PL)
bonf_2 = 1e-3/nrow(Thin_LB_SB_vs_Pi_PL)

#Filter based on bonferroni with alpha = 1e-8
Thin_LB_SB_vs_Pi_PL_filter <- Thin_LB_SB_vs_Pi_PL[Thin_LB_SB_vs_Pi_PL$pvalue < bonf_1,]


#Reshape the maf file 
all_maf_gwas_filter  <- all_maf_filter[all_maf_filter$SNP_ID %in% Thin_LB_SB_vs_Pi_PL_filter$SNP_ID,]
all_maf_gwas_filter  <- all_maf_gwas_filter[all_maf_gwas_filter$Lake_Morph != "Thi_LB_SB" &
                                              all_maf_gwas_filter$Lake_Morph != "Thi_Pi_PL",  ]

#Save the filtered file for future use
write.table(all_maf_gwas_filter, file.path(dir, "Maf_all_pops_filtered_based_BvsP.txt"), col.names = T, row.names = F, quote = F)

```

#Start from this step when the above has been run ones
```{r }
all_maf_gwas_filter <- read.table( file.path(dir, "Maf_all_pops_filtered_based_BvsP.txt"), header = T)
```

## Step 3: Prepare data and select the most common refferense allele in SB morph
```{r }
library(tidyr)
library(reshape2)

# Reshape the data frame
all_maf_filter_matrix <- dcast(all_maf_gwas_filter, SNP_ID ~ Lake_Morph, value.var = "knownEM")


#Add thre refference column based on one benthic morph, I am using ThiSB
all_maf_filter_matrix <- 
  all_maf_filter_matrix %>%
  mutate(Benthic_ref = case_when(ThiSB >= 0.5 ~ 'T',
                                 ThiSB < 0.5 ~ 'F' ) )

#write a funtion to select #Add thre refference column based on one benthic morph, I am using Thi_SB
flip_columns <- function(data, reference_col, test_cols) {
  new_data <- data  # Create a copy of the original data frame
  
  for (col in test_cols) {
    # Create a new column with modified values based on the reference column
    new_col_name <- paste0("flip_", col)
    new_data[[new_col_name]] <- ifelse(data[[reference_col]] == "T", 1 - as.numeric(data[[col]]), data[[col]])
  }
  
  return(new_data)
}

# Apply the function to the data frame with the selected snps
names(all_maf_filter_matrix)

# Specify columns to modify based on the reference column
test_columns_to_flip <- names(all_maf_filter_matrix[2:11])
reference_col <- "Benthic_ref"

# Apply the function
all_maf_filter_matrix_flipped <- flip_columns(all_maf_filter_matrix, reference_col, test_columns_to_flip)


all_maf_filter_matrix_flipped <- all_maf_filter_matrix_flipped[!grepl("unloc", all_maf_filter_matrix_flipped$SNP_ID), ]



#Separated SNP ID column 
all_maf_filter_matrix_flipped_sep <- 
  all_maf_filter_matrix_flipped %>% 
  separate(SNP_ID, into = c('CHR', 'POS'), sep = "_")


```

## Step 4: Select haplotypes - just for visualisation purposes
```{r }
# Define the maximum distance (in base pairs) between SNPs within the same haplotype
max_distance <- 50000

# Load dplyr for data manipulation
library(dplyr)

# Process data to create haplotype groups
haplotypes <- all_maf_filter_matrix_flipped_sep %>%
  mutate(POS = as.numeric(POS)) %>%  # Ensure POS is numeric
  arrange(CHR, POS) %>%              # Order by chromosome and position
  group_by(CHR) %>%                  # Group by chromosome
  mutate(
    Haplotype = cumsum(c(0, diff(POS) > max_distance))  # New haplotype ID when distance exceeds max_distance
  ) %>%
  ungroup()

NumberOfSNP <- 
  haplotypes[, c(1,2,24)] %>%
  group_by(CHR, Haplotype) %>%
  summarise(n_snps = n(), .groups = "drop") %>%
  ungroup()

#Select how many SNPs to cosider within a haplotype
NumberOfSNPs_perHap <- NumberOfSNP[NumberOfSNP$n_snps > 4,]

```




## Step 5: Heatmap preparation
```{r }
all_maf_filter_matrix_flipped_selected <- 
  haplotypes %>%
  semi_join(NumberOfSNPs_perHap, by = c("CHR", "Haplotype"))

#Check the number of crhromosomes
n_distinct(all_maf_filter_matrix_flipped_selected$CHR)


#Columns of interest
columns <- c("CHR", "POS","flip_Myv_pca_DB",  "flip_Myv_pca_Pi",  "flip_SirDB", 
             "flip_SirPL",  "flip_ThiLB",  "flip_ThiPi",  "flip_ThiPL", 
             "flip_ThiSB",  "flip_Van_DB", "flip_Van_PL", "Haplotype" )

#Select only flipped columns
all_maf_filter_matrix_flipped_sel_only <- 
  all_maf_filter_matrix_flipped_selected[,columns ]


#Reshape it back to SNp in columns and Pops in rows
#first make longer
flipped_n <- c("flip_Myv_pca_DB",  "flip_Myv_pca_Pi",  "flip_SirDB", 
               "flip_SirPL",  "flip_ThiLB",  "flip_ThiPi",  "flip_ThiPL", 
               "flip_ThiSB",  "flip_Van_DB", "flip_Van_PL")

long_matrix <- 
  pivot_longer(all_maf_filter_matrix_flipped_sel_only, 
               cols = flipped_n,
               names_to = 'Morphs',
               values_to = "MAF"
)
#Transform matrix into chrnames_snppos colnames
long_matrix$NAME_SNP <- paste(long_matrix$CHR, long_matrix$Haplotype,long_matrix$POS,  sep = '_' )
#[1] "1"  "10" "11" "12" "13" "14" "15" "16" "17" "19"
#[11] "2"  "21" "22" "23" "24" "25" "26" "27" "28" "29"
#[21] "3"  "30" "31" "32" "33" "34" "37" "40" "5"  "6" 
#[31] "7"  "8"  "9" 

#Remove information with inversions 
# Define Positions of interest
Positions <- list(
  chr1 = c(16300000, 18600000),
  chr1_2 = c(19500000, 22200000),
  chr3 = c(33500000, 35800000),
  chr3_2 = c(37350000, 40600000),
  chr8 = c(29000000, 29950000),
  chr9 = c(38400000, 40800000),
  chr14 = c(6300000, 6900000),
  chr40 = c(16250000, 17000000)
)
# Calculate the size of each region
region_sizes <- sapply(Positions, function(pos) {
  size <- pos[2] - pos[1]  # Calculate the region size
  return(size)
})

# Display the sizes for each chromosome region


# Filter out rows that are within specified inversion ranges
filtered_long_matrix <- long_matrix[!(
  (long_matrix$CHR == 1 & long_matrix$POS >= Positions$chr1[1] & long_matrix$POS <= Positions$chr1[2]) |
  (long_matrix$CHR == 1 & long_matrix$POS >= Positions$chr1_2[1] & long_matrix$POS <= Positions$chr1_2[2]) |
  (long_matrix$CHR == 3 & long_matrix$POS >= Positions$chr3[1] & long_matrix$POS <= Positions$chr3[2]) |
  (long_matrix$CHR == 3 & long_matrix$POS >= Positions$chr3_2[1] & long_matrix$POS <= Positions$chr3_2[2]) |
  (long_matrix$CHR == 8 & long_matrix$POS >= Positions$chr8[1] & long_matrix$POS <= Positions$chr8[2]) |
  (long_matrix$CHR == 9 & long_matrix$POS >= Positions$chr9[1] & long_matrix$POS <= Positions$chr9[2]) |
  (long_matrix$CHR == 14 & long_matrix$POS >= Positions$chr14[1] & long_matrix$POS <= Positions$chr14[2]) |
  (long_matrix$CHR == 40 & long_matrix$POS >= Positions$chr40[1] & long_matrix$POS <= Positions$chr40[2])
), ]

#Get the regions min and max positions


#Get the min and max snp positions per each chromosome
positions <- filtered_long_matrix %>%
  group_by(CHR, Haplotype) %>%
  summarise(min_pos= min(POS),
            max_pos = max(POS),
            size = max(POS) - min(POS)
  )

#Numeric chrimosome
positions$CHR <- as.numeric(positions$CHR)

# Convert positions to megabases and keep only one digit after the decimal point
positions$min_pos <- as.numeric(sprintf("%.2f", as.numeric(positions$min_pos) / 1000000))
positions$max_pos <- as.numeric(sprintf("%.2f", as.numeric(positions$max_pos) / 1000000))
positions <- positions %>% arrange(.[[1]])

#Make morph in row, snp in columns
long_matrix_colnamesCHR <- 
  dcast(filtered_long_matrix, Morphs ~ NAME_SNP, value.var = 'MAF' )



#Add sample information 
long_matrix_colnamesCHR_info <- 
  long_matrix_colnamesCHR %>%  
  mutate(Lake = case_when(startsWith(Morphs, "flip_Myv") ~ "Mývatn",
                          startsWith(Morphs, "flip_Thi") ~ "Thingvallavatn", 
                          startsWith(Morphs, "flip_Sir") ~ "Sirdalsvatnet",
                          startsWith(Morphs, "flip_Van") ~ "Vangsvatnet" ),
         Morph = case_when(endsWith(Morphs, "_pca_DB") ~ "SB",
                           endsWith(Morphs,  "ThiPi") ~ "Pi",
                           endsWith(Morphs, "ThiPL") ~ "PL",
                           endsWith(Morphs, "ThiLB") ~ "LB",
                           endsWith(Morphs, "ThiSB") ~ "SB",
                           endsWith(Morphs, 'pca_Pi') ~ "LG",
                          
                           endsWith(Morphs, 'SirDB') ~ "DB",
                           endsWith(Morphs, 'SirPL') ~ "LP",
                           endsWith(Morphs, 'Van_PL') ~ "LP",
                           endsWith(Morphs, 'Van_DB') ~ "DB"
                           
                           ) 
  )

```


## Step 6: Heatmap for all chromosome 
```{r}
data <- long_matrix_colnamesCHR_info

#Order morphs and lakes
level.order = c( 
  "flip_ThiSB", "flip_ThiLB",
  "flip_ThiPL","flip_ThiPi" ,
  "flip_Myv_pca_DB",  "flip_Myv_pca_Pi",
  "flip_SirDB", "flip_SirPL",  
  "flip_Van_DB", "flip_Van_PL"
  
)
data$Morphs <- factor(data$Morphs, levels = level.order)


# Order the data frame based on the factor levels
data <- data[order(data$Morphs), ]

#Prepare matrix
datamatrix <- as.matrix(data[,c(2:(ncol(data)-2)) ]) #2 last column are Morph and lake, flip

#check before running
row.names(datamatrix) <- data$Morphs
datamatrix_ordered <- datamatrix

colnames(datamatrix_ordered) <- sub("^(\\d+_\\d+)_.*", "\\1", colnames(datamatrix_ordered))

#Set up colors
color_list <- list(
   Morph =  c( 
    "DB" = "blue",
    "LP" = "gold",
    'LB' = 'darkslateblue',
    'SB' = 'deepskyblue3',
    'Pi' = 'coral3',
    'PL' = 'chartreuse3',
    "LG" = "darkred"))


```


#Flipped matrix
```{r}
# Transpose the data matrix
datamatrix_ordered <- t(datamatrix)

# Update row and column annotations after transposition
# Now "row_split" is used for chromosomes (as it was in `column_split`)
# and "col_split" is used for lakes (as it was in `row_split`)
col_split <- factor(data$Lake, levels = c("Thingvallavatn","Mývatn",'Sirdalsvatnet',  'Vangsvatnet'))

column_data <- data.frame(
  Lake = c("Thingvallavatn","Mývatn",'Sirdalsvatnet',  'Vangsvatnet'),
  N = c("111", "52", "60", "60") )
column_title <- paste0(column_data$Lake, "\n", "n = ", column_data$N )

# Update row annotations for allele frequencies per chromosome
row_annotation <- factor(positions$CHR[matched_idx])

# Flip the row and column annotations accordingly
# Chromosome and haplotype annotations are now assigned to rows after transposing
row_annot <- rowAnnotation(
  Region = haplotype_annotation,
  col = list(Region = haplotype_colors),
  show_annotation_name = F,
  annotation_name_gp = gpar(fontsize = 8),
  annotation_legend_param = list(
    title_gp = gpar(fontsize = 16),
    labels_gp = gpar(fontsize = 16),
    ncol = 2
  ),
  show_legend = FALSE
)

# Column annotation for lakes
#Order
level.order = c( 
  "flip_ThiSB", "flip_ThiLB",
  "flip_ThiPL","flip_ThiPi" ,
  "flip_Myv_pca_DB",  "flip_Myv_pca_Pi",
  "flip_SirDB", "flip_SirPL",  
  "flip_Van_DB", "flip_Van_PL"
  
)
data$Morphs <- factor(data$Morphs, levels = level.order)
col_order_levels <- unique(data$Morphs)  # Specify the exact column order here


col_annot <- HeatmapAnnotation(
  Morph = data$Morph,
  col = color_list,
  show_annotation_name = FALSE,
  annotation_legend_param = list(
    title_gp = gpar(fontsize = 22), 
    labels_gp = gpar(fontsize = 22)
  ),
  annotation_name_gp = gpar(fontsize = 22)
)

# Plot heatmap with updated row and column orientations
heat_map <- 
  Heatmap(
  datamatrix_ordered,
  col = c("steelblue1", "yellow", "red4"),

  row_order = rownames(datamatrix_ordered),
  column_order = colnames(datamatrix_ordered),
  
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_split = row_annotation,
  column_split = col_split,
  border = FALSE,
  column_title = column_title,
  #row_title = row_title,
  row_title_gp = gpar(fontsize = 12),
  column_title_gp = gpar(fontsize = 20),
  show_column_names = FALSE,
  show_row_names = FALSE,
  show_row_dend = FALSE,
  show_column_dend = FALSE,
  left_annotation = row_annot,
  top_annotation = col_annot,
  row_title_rot = 0,
  heatmap_legend_param = list(
    title = "AF",
    title_gp = gpar(fontsize = 20),
    labels_gp = gpar(fontsize = 20),
    legend_width = unit(2, "cm")
  ),
  column_title_rot = 0
)

pdf(file="/Users/rabu0002/Desktop/Comp_UU/Manuscript/Sumbition/Supplementary_Figures/Supplementary_Fig.7.pdf",
    width = 15, height = 20)
draw(heat_map, merge_legend = TRUE)   
     #heatmap_legend_side = "bottom", 
     #annotation_legend_side = "bottom") 

dev.off()

```


