# Allele frequency of diagnostic SNPs outside the putative inversions selected based on the small/large benthivorous contrast of Arctic charr morphs from Thingvallavatn. 
## Step 1: Load MAF file 
```{r , include= FALSE}
dir = '~/Desktop/Comp_UU/REF_SalAlp_UK/MAF'

#Start
library(tidyverse)
library(dplyr)
library(ggplot2)
library(ggpubr)
options(scipen = 999)
library(scales)
library(ComplexHeatmap)
library(purrr)
library(psych)


#This to prepare Lake_Morph for the heatmap


#Read MAF for all pops
all_maf_filter <- read.table(file.path(dir, "Maf_all_pops.txt"), header = T)
all_maf_filter$SNP_ID <- paste(all_maf_filter$chromo,all_maf_filter$position, sep = "_")

#Add morph info
all_maf_filter <- 
  all_maf_filter %>%  
  mutate(Lake = case_when(substr(Lake_Morph, 1, 3) == "Myv" ~ "Mývatn",
                          substr(Lake_Morph, 1, 3) == "Thi" ~ "Thingvallavatn", 
                          substr(Lake_Morph, 1, 3) == "Sir" ~ "Sirdalsvatnet",
                          substr(Lake_Morph, 1, 3) == "Van"~ "Vangsvatnet" )
  )

```

## Step 2: Select outlier SNPs based on LB vs SB contrast
```{r }
Thin_LB_vs_SB <- read.table('~/Desktop/Comp_UU/REF_SalAlp_UK/GWAS/Data/Thi_DB_vs_LB.pvalue', header = T)
Thin_LB_vs_SB$SNP_ID <- paste(Thin_LB_vs_SB$Chromosome,Thin_LB_vs_SB$Position, sep = "_")

bonf = 1e-8/nrow(Thin_LB_vs_SB)

#Filter based on bonferroni
Thin_LB_vs_SB_filter <- Thin_LB_vs_SB[Thin_LB_vs_SB$pvalue < bonf,] #< because its not -log 10 


#Reshape the maf file 
all_maf_gwas_filter  <- all_maf_filter[all_maf_filter$SNP_ID %in% Thin_LB_vs_SB_filter$SNP_ID,]
all_maf_gwas_filter  <- all_maf_gwas_filter[all_maf_gwas_filter$Lake_Morph != "Thi_LB_SB" &
                                            all_maf_gwas_filter$Lake_Morph != "Thi_Pi_PL",  ]

#Save the filtered file for future use
write.table(all_maf_gwas_filter, file.path(dir, "Maf_all_pops_filtered_based_LBvsSB.txt"), col.names = T, row.names = F, quote = F)

```

## Start from this step when the above has been run ones
```{r }
all_maf_gwas_filter <- read.table( file.path(dir, "Maf_all_pops_filtered_based_BvsP.txt"), header = T)
```



## Step 3: Prepare data and select the most common refferense allele in on one of the benthic morphs (SB)
```{r }
library(tidyr)
library(reshape2)

# Reshape the data frame
all_maf_filter_matrix <- dcast(all_maf_gwas_filter, SNP_ID ~ Lake_Morph, value.var = "knownEM")


#Add thre refference column based on one benthic morph, I am using ThiSB
all_maf_filter_matrix <- 
  all_maf_filter_matrix %>%
  mutate(Benthic_ref = case_when(ThiSB >= 0.5 ~ 'T',
                                 ThiSB < 0.5 ~ 'F' ) )

#write a funtion to select #Add thre refference column based on one benthic morph, I am using Thi_SB
flip_columns <- function(data, reference_col, test_cols) {
  new_data <- data  # Create a copy of the original data frame
  
  for (col in test_cols) {
    # Create a new column with modified values based on the reference column
    new_col_name <- paste0("flip_", col)
    new_data[[new_col_name]] <- ifelse(data[[reference_col]] == "T", 1 - as.numeric(data[[col]]), data[[col]])
  }
  
  return(new_data)
}

# Apply the function to the data frame with the selected snps

#all_maf_filter_matrix <- all_maf_filter_matrix[, -c("Thi_LB_SB","Thi_Pi_PL")]

# Specify columns to modify based on the reference column
test_columns_to_flip <- names(all_maf_filter_matrix[2:11])
reference_col <- "Benthic_ref"

# Apply the function
all_maf_filter_matrix_flipped <- flip_columns(all_maf_filter_matrix, reference_col, test_columns_to_flip)

## ## ## ## ## ## ## ## ## ## ## ## ## ## ## #Check for NA critical step## ## ## ## ## ## ## ## ## #
# Replace columns with NA to 0
all_maf_filter_matrix_flipped_nato0 <- all_maf_filter_matrix_flipped
all_maf_filter_matrix_flipped_nato0[is.na(all_maf_filter_matrix_flipped_nato0)] <- 0

# Remove columns with NA 
all_maf_filter_matrix_flipped_noNA <- all_maf_filter_matrix_flipped[ , colSums(is.na(all_maf_filter_matrix_flipped))==0]

## ## ## ## ## ## ## ## ## ## ## ## ## ## ## #Check for NA critical step## ## ## ## ## ## ## ## ## #

#Separated SNP ID column 
all_maf_filter_matrix_flipped_sep <- 
  all_maf_filter_matrix_flipped %>% 
  separate(SNP_ID, into = c('CHR', 'POS'), sep = "_")


```



## Step 4: Select haplotypes based on conditions
```{r }
# Define the maximum distance (in base pairs) between SNPs within the same haplotype
max_distance <- 50000 #just for visualising purposes otherwise all regions are displaced

# Load dplyr for data manipulation
library(dplyr)

# Process data to create haplotype groups
haplotypes <- all_maf_filter_matrix_flipped_sep %>%
  mutate(POS = as.numeric(POS)) %>%  # Ensure POS is numeric
  arrange(CHR, POS) %>%              # Order by chromosome and position
  group_by(CHR) %>%                  # Group by chromosome
  mutate(
    Haplotype = cumsum(c(0, diff(POS) > max_distance))  # New haplotype ID when distance exceeds max_distance
  ) %>%
  ungroup()

NumberOfSNP <- 
  haplotypes[, c(1,2,24)] %>%
  group_by(CHR, Haplotype) %>%
  summarise(n_snps = n(), .groups = "drop") %>%
  ungroup()

#Select how many SNPs to cosider within a haplotype
NumberOfSNPs_perHap <- NumberOfSNP[NumberOfSNP$n_snps > 4,]

```



## Step 5: Heatmap preparation
```{r }
all_maf_filter_matrix_flipped_selected <- 
  haplotypes %>%
  semi_join(NumberOfSNPs_perHap, by = c("CHR", "Haplotype"))

#Check the number of crhromosomes
n_distinct(all_maf_filter_matrix_flipped_selected$CHR)


#Columns of interest
columns <- c("CHR", "POS","flip_Myv_pca_DB",  "flip_Myv_pca_Pi",  "flip_SirDB", 
             "flip_SirPL",  "flip_ThiLB",  "flip_ThiPi",  "flip_ThiPL", 
             "flip_ThiSB",  "flip_Van_DB", "flip_Van_PL", "Haplotype" )

#Select only flipped columns
all_maf_filter_matrix_flipped_sel_only <- 
  all_maf_filter_matrix_flipped_selected[,columns ]


#Reshape it back to SNp in columns and Pops in rows
#first make longer
flipped_n <- c("flip_Myv_pca_DB",  "flip_Myv_pca_Pi",  "flip_SirDB", 
               "flip_SirPL",  "flip_ThiLB",  "flip_ThiPi",  "flip_ThiPL", 
               "flip_ThiSB",  "flip_Van_DB", "flip_Van_PL")

long_matrix <- 
  pivot_longer(all_maf_filter_matrix_flipped_sel_only, 
               cols = flipped_n,
               names_to = 'Morphs',
               values_to = "MAF"
)



#Transform matrix into chrnames_snppos colnames
long_matrix$NAME_SNP <- paste(long_matrix$CHR, long_matrix$Haplotype,long_matrix$POS,  sep = '_' )


# Define Positions of interest (in base pairs)
Positions <- list(
  'chr4' = c(75250000, 76130000),
  'chr5' = c(22300000, 22750000),
  'chr9' = c(61300000, 62110000),
  'chr17' = c(32450000, 33200000)
)

# Calculate the size of each region
region_sizes <- sapply(Positions, function(pos) {
  size <- pos[2] - pos[1]  # Calculate the region size
  return(size)
})


# Filter out rows that are within specified inversion ranges
filtered_long_matrix <- long_matrix[!(
  (long_matrix$CHR == 4 & long_matrix$POS >= Positions$chr4[1] & long_matrix$POS <= Positions$chr4[2]) |
  (long_matrix$CHR == 5 & long_matrix$POS >= Positions$chr5[1] & long_matrix$POS <= Positions$chr5[2]) |
  (long_matrix$CHR == 9 & long_matrix$POS >= Positions$chr9[1] & long_matrix$POS <= Positions$chr9[2]) |
  (long_matrix$CHR == 17 & long_matrix$POS >= Positions$chr17[1] & long_matrix$POS <= Positions$chr17[2])
), ]


#Get the min and max snp positions per each chromosome
positions <- filtered_long_matrix %>%
  group_by(CHR, Haplotype) %>%
  summarise(min_pos= min(POS),
            max_pos = max(POS)
  )

#Numeric chromosome
positions$CHR <- as.numeric(positions$CHR)

# Convert positions to megabases and keep only one digit after the decimal point
positions$min_pos <- as.numeric(sprintf("%.2f", as.numeric(positions$min_pos) / 1000000))
positions$max_pos <- as.numeric(sprintf("%.2f", as.numeric(positions$max_pos) / 1000000))
positions <- positions %>% arrange(.[[1]])


#Make morph in row, snp in columns
long_matrix_colnamesCHR <- 
  dcast(filtered_long_matrix, Morphs ~ NAME_SNP, value.var = 'MAF' )

#Add sample information 
long_matrix_colnamesCHR_info <- 
  long_matrix_colnamesCHR %>%  
  mutate(Lake = case_when(startsWith(Morphs, "flip_Myv") ~ "Mývatn",
                          startsWith(Morphs, "flip_Thi") ~ "Thingvallavatn", 
                          startsWith(Morphs, "flip_Sir") ~ "Sirdalsvatnet",
                          startsWith(Morphs, "flip_Van") ~ "Vangsvatnet" ),
         Morph = case_when(endsWith(Morphs, "_pca_DB") ~ "SB",
                           endsWith(Morphs,  "ThiPi") ~ "Pi",
                           endsWith(Morphs, "ThiPL") ~ "PL",
                           endsWith(Morphs, "ThiLB") ~ "LB",
                           endsWith(Morphs, "ThiSB") ~ "SB",
                           endsWith(Morphs, 'pca_Pi') ~ "LG",
                         
                           endsWith(Morphs, 'SirDB') ~ 'DB',
                           endsWith(Morphs, 'SirPL') ~ "LP",
                          
                           endsWith(Morphs, 'Van_PL') ~ "LP",
                           endsWith(Morphs, 'Van_DB') ~ "DB"
                           
                           ) 
  )

```


## Step 5: Heatmap plot
```{r}
data <- long_matrix_colnamesCHR_info

#Order
level.order = c( 
  "flip_ThiSB", "flip_ThiLB",
  "flip_ThiPL","flip_ThiPi" ,
  "flip_Myv_pca_DB",  "flip_Myv_pca_Pi",
  "flip_SirDB", "flip_SirPL",  
  "flip_Van_DB", "flip_Van_PL"
  
)
data$Morphs <- factor(data$Morphs, levels = level.order)


# Order the data frame based on the factor levels
data <- data[order(data$Morphs), ]

#Prepare matrix
datamatrix <- as.matrix(data[,c(2:(ncol(data)-2)) ]) #2 last colums are Morph and lake, flip

#check before running
row.names(datamatrix) <- data$Morphs
datamatrix_ordered <- datamatrix
#colnames(datamatrix_ordered) <- gsub('_.*', "", colnames(datamatrix_ordered) )
colnames(datamatrix_ordered) <- sub("^(\\d+_\\d+)_.*", "\\1", colnames(datamatrix_ordered))


# Plot heatmap settings-------------------------------------------------
#Set up colors
color_list <- list(
   Morph =  c( 
    "DB" = "blue",
    "LP" = "gold",
    'LB' = 'darkslateblue',
    'SB' = 'deepskyblue3',
    'Pi' = 'coral3',
    'PL' = 'chartreuse3',
    "LG" = "darkred"))


#Create annotation for rows
annot <- 
  rowAnnotation(
    Morph = data$Morph,
    col = color_list,
   show_annotation_name = FALSE,
    annotation_legend_param = list(
      title_gp = gpar(fontsize = 20), 
      labels_gp = gpar(fontsize = 20)
    ),
    annotation_name_gp= gpar(fontsize = 20)
  )

#Add annotation column to positions
positions$Chr_pos <- paste("Scaffold", positions$CHR, ":", positions$min_pos, "-", 
                           positions$max_pos, sep = " ")
positions$Chr_hap <- paste(positions$CHR, positions$Haplotype, sep = "_")


#Annotation splits
row_split <- factor(data$Lake, levels = c("Thingvallavatn","Mývatn",'Sirdalsvatnet',  'Vangsvatnet'))
row_data <- data.frame(
  Lake = c("Thingvallavatn","Mývatn",'Sirdalsvatnet',  'Vangsvatnet'),
  N = c("111", "52", "60", "60") )
row_title <- paste0(row_data$Lake, "\n", "n = ", row_data$N )

# Define two colors to alternate for haplotypes
alternating_haplotype_colors <- c("brown4", "orange2")


# Ensure `positions$Chr_hap` is correctly formatted
positions$Chr_hap <- paste(positions$CHR, positions$Haplotype, sep = "_") 

# Match datamatrix columns with positions based on Chr_hap
matched_idx <- match(colnames(datamatrix_ordered), positions$Chr_hap)

# Check for NAs and report any missing values
if (any(is.na(matched_idx))) {
  warning("Some columns in datamatrix_ordered have no match in positions$Chr_hap.")
  print(colnames(datamatrix_ordered)[is.na(matched_idx)])
}



# Extract chromosome and haplotype annotations
chromosome_annotation <- factor(positions$CHR[matched_idx])
haplotype_annotation <- factor(positions$Chr_pos[matched_idx], levels = positions$Chr_pos)


# Count unique haplotypes (assumes `haplotype_annotation` contains all unique haplotype labels)
num_haplotypes <- length(unique(haplotype_annotation))

# Create a named color vector with alternating colors for each unique haplotype
# This replicates the two colors across all unique haplotype labels
haplotype_colors <- setNames(
  rep(alternating_haplotype_colors, length.out = num_haplotypes),
  unique(haplotype_annotation)
)


# Check lengths to ensure they match the number of columns in datamatrix_ordered
if (length(chromosome_annotation) != ncol(datamatrix_ordered) || 
    length(haplotype_annotation) != ncol(datamatrix_ordered)) {
  stop("Mismatch in number of annotations and columns in datamatrix_ordered.")
}


# Continue with Heatmap creation
col_annot <- HeatmapAnnotation(

  Region = haplotype_annotation,  # Use the factor based on your haplotype names
  col = list(Region = haplotype_colors),  # Apply alternating colors
  show_annotation_name = F,
  annotation_name_gp = gpar(fontsize = 22),
  annotation_legend_param = list(
      title_gp = gpar(fontsize = 20), 
      labels_gp = gpar(fontsize =20),
      nrow = 1,
      legend_direction = "horizontal",
      title = NULL
    )
)


heat_map <- Heatmap(datamatrix,
        col = c("steelblue1", "yellow", "red4"),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_order = rownames(datamatrix),
  column_order = colnames(datamatrix),
  row_split = row_split,
  column_split = chromosome_annotation,  # Split columns by chromosome
  border = FALSE,
  row_title = row_title,
  row_title_gp = gpar(fontsize = 20),
  column_title_gp = gpar(fontsize = 20),
  show_column_names = FALSE,
  show_row_names = FALSE,
  show_row_dend = FALSE,
  show_column_dend = FALSE,
  #column_title = NULL,
  left_annotation = annot,
  top_annotation = col_annot,  # Chromosome and haplotype annotations
  
  row_title_rot = 90,
  heatmap_legend_param = list(
    title = "AF",
    title_gp = gpar(fontsize = 20),
    labels_gp = gpar(fontsize = 20),
    legend_width = unit(1, "cm")
  ),
  #column_title_rot = 0
)


pdf(file="/Users/rabu0002/Desktop/Comp_UU/Manuscript/Sumbition/Supplementary_Figures/Supplementary_Fig.6.pdf", width = 14, height = 10)
draw(heat_map,  
     annotation_legend_side = "bottom"  # Place heatmap legend at the bottom
     ) 
dev.off()
```
