
## Step 1: Estimate LD within the region
```{r }
#!/bin/bash -l

#SBATCH -A naiss2024-5-277
#SBATCH -p core -n 10
#SBATCH -t 02-00:00:00
#SBATCH --array=1-2
#SBATCH -J LD_MyvThin
#SBATCH -e LD_MyvThin_%A_%a.err
#SBATCH -o LD_MyvThin_%A_%a.out
#SBATCH --mail-type=all
#SBATCH --mail-user=khrystyna.kurta@slu.se

#Load modules
module load bioinfo-tools
module load ANGSD/0.933
module load ngsLD/1.1.1

#Directorys
BASEDIR=/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/GL_MAF_GENO
BEAGLE_MAIN=$BASEDIR/Myv_CharrLowPass_GATKMethod_MinMAF0.05_all_chr.beagle.gz

OUT_DIR=/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/GL_MAF_GENO_SHARED
cd $OUT_DIR

#Chr34
{
    zcat $BEAGLE_MAIN | head -n 1  # Extract the header
    zcat $BEAGLE_MAIN | awk -F'[_\t]' '$1 == "34" && $2 >= 18300000 && $2 <= 18500000' 
} | gzip > Myv_CharrLowPass_GATKMethod_MinMAF0.05_shared_34chr_18.3_18.5Mb.beagle.gz

ls Myv_CharrLowPass_GATKMethod_MinMAF0.05_shared_34chr_18.3_18.5Mb.beagle.gz Thin_CharrLowPass_GATKMethod_MinMAF0.05_shared_34chr_18.3_18.5Mb.beagle.gz > shared_b.list

BEAGLE=$(cat shared_b.list | sed -n ${SLURM_ARRAY_TASK_ID}p)

#Names
NAME=${BEAGLE/.beagle.gz/}

#A beagle formatted genotype likelihood file generated from ANGSD (-doGlf 2) can be inputted into ngsLD after the header row and the first three columns (i.e. positions, major allele, minor allele) are removed.

# Preparing pos file by splitting chromosome and position
zcat $BEAGLE | \
    tail -n +2 |
    awk '{split($1, a, "_"); print a[1] "\t" a[2]}' | \
    gzip > ${NAME}.pos.gz
    

# Counting number of sites
N_SITES=$(zcat ${NAME}.pos.gz | wc -l)

zcat $BEAGLE | wc -l
zcat ${NAME}.pos.gz | wc -l
echo $N_SITES

# Print the result information
echo "For BEAGLE ${NAME}_subsampled.beagle.gz  positions ${NAME}.pos.gz with number of sites $N_SITES"

#Run LD
ngsLD \
--geno $BEAGLE \
--pos ${NAME}.pos.gz \
--probs \
--n_ind 54 \
--n_sites $N_SITES \
--max_kb_dist 0 \
--n_threads 8 \
--out ${NAME}.ld
#--min_maf 0.1


#Visualize LD blocks in R
```


## Step 2: Plot Zoom-in profile of GWC and linkage disequilibrium on scaffold 34 (18.36-18.45 Mb). 
```{r }
# 1. Load packages
library(data.table)
library(ggplot2)
library(scales)

dir="~/Desktop/Comp_UU/REF_SalAlp_UK/GWAS/Data/"
ld_dir="~/Desktop/Comp_UU/REF_SalAlp_UK/LD/"


# 2. Read GWAS files
thin_gwas <- fread(file.path(dir, 'Thi_DB_LB_vs_PL_Pi.pvalue'))
myv_gwas <- fread(file.path(dir, 'Myv_PCA_SB_PL.pvalue'))

# 3. Read LD files
thin_ld <- fread(file.path(ld_dir, "Thin_GATKMethod_MinMAF0.05_shared_34chr_18.3_18.5Mb.ld"), select = 1:4)
myv_ld  <- fread(file.path(ld_dir, "Myv_CharrLowPass_GATKMethod_MinMAF0.05_shared_34chr_18.3_18.5Mb.ld"), select = 1:4)

# 4. Define region
chr <- 34
minn <- 18360000
maxx <- 18450000



# 5. Prepare LD file
file <- myv_ld  # use your selected LD dataset
setnames(file, c("Pos1", "Pos2", "Distance", "R2_Pearson"))

file[, R2_Pearson := as.numeric(R2_Pearson)]
file <- file[!is.na(R2_Pearson)]

file[, Chr := as.numeric(sub(":.*", "", Pos1))]
file[, Pos1 := as.numeric(sub(".*:", "", Pos1))]
file[, Pos2 := as.numeric(sub(".*:", "", Pos2))]
file[, Pos1_Mb := round(Pos1 / 1e6, 3)]
file[, Pos2_Mb := round(Pos2 / 1e6, 3)]


# 6. Adjust axis based on LD content
axis_min <- min(file$Pos2)
axis_max <- max(file$Pos2)

# 7. Prepare GWAS data
data <- myv_gwas
data[, Position := as.numeric(Position)]
data[, Position_Mb := Position / 1e6]

# Bonferroni thresholds
bonf_1 <- -log10(0.001 / nrow(data))
bonf_2 <- -log10(1e-8 / nrow(data))

# 8. Find top SNP within the region
top_snp <- data[Chromosome == chr & Position >= axis_min & Position <= axis_max][which.min(pvalue)]
top_pos <- top_snp$Position

ld_with_top <- file[Pos1 == top_pos | Pos2 == top_pos,
                    .(OtherPos = ifelse(Pos1 == top_pos, Pos2, Pos1), R2_Pearson)]

ld_with_top[, Position := OtherPos][, OtherPos := NULL]

# 9. Filter GWAS data for region
gwas_region <- data[Chromosome == chr & Position >= axis_min & Position <= axis_max]


# Merge GWAS with LD R² for highlight
highlight_region <- merge(gwas_region, ld_with_top, by = "Position", all.x = TRUE)


# 10. GWAS Manhattan plot with LD coloring
plot_gwas_myv <- ggplot(gwas_region) +
  geom_point(aes(x = Position_Mb, y = -log10(pvalue)), size = 1, color = "black") +
  geom_point(data = highlight_region, aes(x = Position_Mb, y = -log10(pvalue), color = R2_Pearson), size = 1.2) +
  scale_color_gradientn(colors = c("yellow2", "red", "darkblue"), 
                        na.value = "darkgrey", name = expression(R^2)) +
  geom_hline(yintercept = bonf_1, linetype = "dashed", color = "red") +
  geom_hline(yintercept = bonf_2, linetype = "dashed", color = "blue") +
  scale_y_continuous(labels = label_number(accuracy = 0.02),
                     limits = c(0.00, 25.00)) +
  scale_x_continuous(limits = c(axis_min, axis_max) / 1e6,
                     labels = label_number(accuracy = 0.01)) +
  theme_classic(base_size = 16) +
  theme(legend.position = 'right',
        axis.title.x = element_blank(),
        axis.text.x = element_blank()) +
  labs(y = "-log10(P-value)")


# 11. LD Heatmap
my_ld_plot <- 
  ggplot(file, aes(x = Pos1_Mb, y = Pos2_Mb, fill = R2_Pearson)) +
  geom_tile()+ 
  scale_fill_gradientn(colors = c("yellow2", "red", "darkblue"), na.value = "grey", name = expression(R^2)) +
  scale_x_continuous(labels = label_number(accuracy = 0.01),
                     limits = c(axis_min, axis_max) / 1e6) +
  scale_y_continuous(labels = label_number(accuracy = 0.01),
                     limits = c(axis_min, axis_max) / 1e6) +
  labs(x = paste("Scaffold", chr, "position (Mb)"), y = "Position (Mb)") +
  theme_classic(base_size = 16) +
  theme(legend.title = element_text(size = 9),
        legend.text = element_text(size = 9),
        legend.key.size = unit(0.4, "cm"),
        legend.position = 'none')


# 12. LD Stats
sumstat <- 
  highlight_region %>%
  filter(Position >= minn & Position <= maxx) %>%
  summarise(
    min_R = min(R2_Pearson, na.rm = TRUE),
    max_R = max(R2_Pearson, na.rm = TRUE),
    mean = mean(R2_Pearson, na.rm = TRUE),
    sd_R  = sd(R2_Pearson, na.rm = TRUE)
  )

# Thingvallavatn  #               
name="Thingvallavatn, Benthic vs. Pelagic"
data = thin_gwas


# 5. Prepare LD file
file = thin_ld 
setnames(file, c("Pos1", "Pos2", "Distance", "R2_Pearson"))


file[, R2_Pearson := as.numeric(R2_Pearson)]
file <- file[!is.na(R2_Pearson)]

file[, Chr := as.numeric(sub(":.*", "", Pos1))]
file[, Pos1 := as.numeric(sub(".*:", "", Pos1))]
file[, Pos2 := as.numeric(sub(".*:", "", Pos2))]
file[, Pos1_Mb := round(Pos1 / 1e6, 3)]
file[, Pos2_Mb := round(Pos2 / 1e6, 3)]


# 6. Adjust axis based on LD content
axis_min <- min(file$Pos2)
axis_max <- max(file$Pos2)

# 7. Prepare GWAS data
data = thin_gwas
data[, Position := as.numeric(Position)]
data[, Position_Mb := Position / 1e6]

# Bonferroni thresholds
bonf_1 <- -log10(0.001 / nrow(data))
bonf_2 <- -log10(1e-8 / nrow(data))

# 8. Find top SNP within the region
top_snp <- data[Chromosome == chr & Position >= minn & Position <= maxx][which.min(pvalue)]
top_pos <- top_snp$Position

ld_with_top <- file[Pos1 == top_pos | Pos2 == top_pos,
                    .(OtherPos = ifelse(Pos1 == top_pos, Pos2, Pos1), R2_Pearson)]

ld_with_top[, Position := OtherPos][, OtherPos := NULL]

# 9. Filter GWAS data for region
gwas_region <- data[Chromosome == chr & Position >= axis_min & Position <= axis_max]


# Merge GWAS with LD R² for highlight
highlight_region <- merge(gwas_region, ld_with_top, by = "Position", all.x = TRUE)


# 10. GWAS Manhattan plot with LD coloring
plot_gwas_thin <- ggplot(gwas_region) +
  geom_point(aes(x = Position_Mb, y = -log10(pvalue)), size = 1, color = "black") +
  geom_point(data = highlight_region, aes(x = Position_Mb, y = -log10(pvalue), color = R2_Pearson), size = 1.2) +
  scale_color_gradientn(colors = c("yellow2", "red", "darkblue"), 
                        na.value = "darkgrey", name = expression(R^2)) +
  geom_hline(yintercept = bonf_1, linetype = "dashed", color = "red") +
  geom_hline(yintercept = bonf_2, linetype = "dashed", color = "blue") +
  scale_y_continuous(labels = label_number(accuracy = 0.02),
                     limits = c(0.00, 25.00)) +
  scale_x_continuous(limits = c(axis_min, axis_max) / 1e6,
                     labels = label_number(accuracy = 0.01)) +
  theme_classic(base_size = 16) +
  theme(legend.position = 'right',
        axis.title.x = element_blank(),
        axis.text.x = element_blank()) +
  labs(y = "-log10(P-value)")

# 11. LD Heatmap
ld_plot_thing <- 
  ggplot(file, aes(x = Pos1_Mb, y = Pos2_Mb, fill = R2_Pearson)) +
  geom_tile()+ 
  scale_fill_gradientn(colors = c("yellow2", "red", "darkblue"), na.value = "grey", name = expression(R^2)) +
  scale_x_continuous(labels = label_number(accuracy = 0.01),
                     limits = c(axis_min, axis_max) / 1e6) +
  scale_y_continuous(labels = label_number(accuracy = 0.01),
                     limits = c(axis_min, axis_max) / 1e6) +
  labs(x = paste("Scaffold", chr, "position (Mb)"), y = "Position (Mb)") +
  theme_classic(base_size = 16) +
  theme(legend.title = element_text(size = 9),
        legend.text = element_text(size = 9),
        legend.key.size = unit(0.4, "cm"),
        legend.position = 'none')


# 12. LD Stats
sumstat <- 
  highlight_region %>%
  filter(Position >= minn & Position <= maxx) %>%
  summarise(
    min_R = min(R2_Pearson, na.rm = TRUE),
    max_R = max(R2_Pearson, na.rm = TRUE),
    mean = mean(R2_Pearson, na.rm = TRUE),
    sd_R  = sd(R2_Pearson, na.rm = TRUE)
  )


#Arrange all plots
ggsave('~/Desktop/Comp_UU/REF_SalAlp_UK/GWAS/GWAS_34chr_zoom.png',ggarrange(
  plot_gwas_myv,
  my_ld_plot, 
  plot_gwas_thin,
  ld_plot_thing,
  nrow = 4, ncol = 1
), width = 6,  height = 8, dpi = 300)

```
