# Functional annotation of gene models
```{r}
#!/bin/bash -l

#SBATCH -A naiss2024-5-277
#SBATCH -p core -n 20
#SBATCH -t 09-00:00:00
#SBATCH -J GA
#SBATCH -e GA_%A_%a.err
#SBATCH -o GA_%A_%a.out
#SBATCH --mail-type=all
#SBATCH --mail-user=khrystyna.kurta@slu.se

#Load modules
module load bioinfo-tools
module load BEDTools/2.31.1
module load AGAT/1.3.2
module load Magic-BLAST/1.7.0
module load blast/2.15.0+
module load InterProScan/5.62-94.0
module load samtools/1.20
module load snpEff/5.2

DIR=/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/GENE_ANNOTATION
cd $DIR


#REF
REFDIR=/proj/snic2020-2-19/private/arctic_charr/assemblies/fSalAlp1.1/Data_Package_fSalAlp1_assembly_20231024

# Update this with the name of the ref fasta file
REF=/proj/snic2020-2-19/private/arctic_charr/assemblies/fSalAlp1.1/Data_Package_fSalAlp1_assembly_20231024/fSalAlp1.1.hap1.cur.20231016.fasta

#GFF
GFF=Saalp_EIv1.0.release.functional_annotation.gff3

#STEP 1:
#Prepare inversion coordinates:
bedtools getfasta -fi $REF -bed Saalp_EIv1.0.release.functional_annotation.gff3 -name -fo genes.fa

#STERP 3:
# Translate CDS to protein sequences
agat_sp_extract_sequences.pl -g $GFF -f genes.fa -t cds -p --merge -o proteins.fa


#STEP 4:
# Download UniProtKB/SwissProt database (version 2021_03) - this should be done ones only
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz
gunzip uniprot_sprot.fasta.gz

#STEP 5:
# Format the database for BLAST
makeblastdb -in uniprot_sprot.fasta -dbtype prot -out uniprot_db


#STEP 6:
#Select chromosome of interest
for chrom in 1 3 4 5 8 9 14 17 34 40; do
    awk -v chrom="$chrom" '/^>/ {p = ($0 ~ "seq_id=" chrom "[^0-9]")} p' proteins.fa > "selected_blastp_chr$chrom.fa"
done


# Run BLASTp 
blastp -query "selected_blastp_chr$chrom.fa" -db uniprot_db -evalue 1e-6 -outfmt 6 -out blast_results_test.tsv -num_threads 8

#STEP 5: This step as blastp also annotates genes
# Run InterProScan for functional annotation - remove stars because interproscan cannot read them
sed 's/*//g' proteins.fa > proteins_clean.fa 

for chrom in 1 3 4 5 8 9 14 17 34 40; do
    awk -v chrom="$chrom" '/^>/ {p = ($0 ~ "seq_id=" chrom "[^0-9]")} p' proteins_clean.fa > "selectedInterPro_chr${chrom}.fa"
done

interproscan.sh -i "selectedInterPro_chr${chrom}.fa" -f tsv -dp -b interproscan_results

#STEP 7: Run SnpEff 
#Validate dummy VCF file with the selected positions for SNPEff 
vcftools --vcf Thi_DB_LB_inv.vcf --out Thi_DB_vs_LB_val.vcf --recode
vcftools --vcf Thi_DB_LB_vs_PL_Pi_inv.vcf --out Thi_DB_LB_vs_PL_Pi_val.vcf --recode

#Run SnpEff
#SnpEff also generates a TXT (tab separated) file having counts of number of variants affecting each transcript and gene. By default, the file name is snpEff_genes.txt, but it can be changed using the -stats command line option.

java -Xmx6g -jar /home/khrys/snpEff/snpEff.jar Saalp_EIv1.0 Thi_DB_vs_LB_val.vcf.recode.vcf -stats Thi_DB_vs_LB_genes.txt > Thi_DB_vs_LB_annotated_snps.vcf

java -Xmx6g -jar /home/khrys/snpEff/snpEff.jar Saalp_EIv1.0 Thi_DB_LB_vs_PL_Pi_val.vcf.recode.vcf -stats Thi_DB_LB_vs_PL_Pi_genes.txt > Thi_DB_LB_vs_PL_Pi_annotated_snps.vcf

```
