
#  Functional annotation of gene models - R
```{r setup, include=FALSE}
library(BiocManager)
require(GenomicRanges)
require(IRanges)
require(rtracklayer)
require(tidyverse)
require(ggpubr)
require(dplyr)
require(grid)
library(xlsx)

## Add functional annotation data from blast and interpro

## Define a directory
dir = "~/Desktop/Comp_UU/REF_SalAlp_UK/GENE_ANNOTATION/RESULTS"

##  List files in the directory
res <- list.files(dir)

##  Separate BLAST and InterPro files
blastp_files <- grep("blast", res, value = TRUE)
interpro_files <- grep("interpro", res, value = TRUE)

##  Initialize a list to store results
blast_results <- list()
interPro_results <- list()

## Set chromosomes of interest 
chrs <- c('chr1','chr3','chr4','chr5','chr8','chr9','chr14','chr17','chr34','chr40')

##  Process files by chromosome
for (chr in chrs) {
  
  ##  Get BLAST and InterPro files for the current chromosome
  blast_file <- grep(paste0(chr, "\\.fa_blast_results\\.tsv$"), blastp_files, value = TRUE)
  interpro_file <- grep(paste0( chr, "\\.fa_interproscan_results\\.tsv.tsv$"), interpro_files, 
                        value = TRUE)

  
  if (length(blast_file) > 0 && length(interpro_file) > 0) {
    blast <- read.delim(file.path(dir, blast_file), header = FALSE, sep = "\t", 
                                stringsAsFactors = FALSE)
    colnames(blast) <- c("query", "subject", "identity", "alignment_length", 
                                 "mismatches", "gap_opens", 
                                 "q_start", "q_end", "s_start", "s_end", "evalue", 
                                 "bit_score")
    
    interpro <- read.delim(file.path(dir,interpro_file), header = F, sep = "\t", 
                                   stringsAsFactors = FALSE)
    
    colnames(interpro) <- c("query", "Sequence_MD5_digest", "Sequence_length", 
                                 "Analysis", "Signature_accession", 
                                 "Signature_description", "sig_sart", "sig_end", "e_value", 
                                 "Status","Date_run", "InterPro_AnnotationAssessin", 
                                 "InterPro_AnnotationDescription")
    
    
    
    ##  Save results for the current chromosome
    blast_results[[paste0(chr)]] <- blast
    interPro_results[[paste0(chr)]] <- interpro
  }
}


##  Add the data frame names as a new column
blast_results_names <- Map(cbind, 
                                Name = names(blast_results), 
                                blast_results)
##  Combine the data frames into a single table
blast_results_all_chr <- do.call(rbind, blast_results_names)


blast_results_all_chr_best_hit <- 
  blast_results_all_chr %>%
  group_by(query, Name) %>%
  reframe(best_hit = subject[which.min(evalue)],
            evalue = evalue[which.min(evalue)],
            identity = identity[which.min(evalue)]
            )

## Get gene ids
blast_results_all_chr_best_hit_geneID <- blast_results_all_chr_best_hit %>%
  separate(query, into = c("GeneID", "Transcript"), sep = "\\.(?=[0-9]+$)", remove = FALSE)

blast_results_all_chr_best_hit_geneID <- blast_results_all_chr_best_hit_geneID[blast_results_all_chr_best_hit_geneID$Transcript ==1,]

blast_results_all_chr_best_hit_geneID$Gene_Symbol <-
sub(".*\\|(.*)_.*", "\\1", blast_results_all_chr_best_hit_geneID$best_hit)

## Any duplicated gene Ids?
any(duplicated(blast_results_all_chr_best_hit_geneID$GeneID))

```


## Get GWAS infromation
```{r }
dir_gwas <- '~/Desktop/Comp_UU/REF_SalAlp_UK/GWAS/Data'

## Load data 
files <- list.files(dir_gwas)

##  Filter files 
desired_files <- grep("pvalue$", files, value = TRUE)

## Look at the file list 
remove_t <- c("Sir_D_vs_N.pvalue", ## all regions differentiated
              "Van_D_vs_N.pvalue", ## no differentiated regions
              "Myv_DB_vs_Pi.pvalue" ## preliminary classified morphs
              )

##  Filter out unwanted strings
filtered_list <- desired_files[!desired_files %in% remove_t]

##  Read the desired files
gwas_list <- list()

for(file in filtered_list) {
  file_content <- read.table(file.path(dir_gwas, file), header = T)
  
  ##  Process the file content as needed
  gwas_list[[file]] <- file_content
}


## Modify names of the data list 
names(gwas_list) <- gsub(".pvalue", '', names(gwas_list) )


## Replace NA with 0
gwas_filt_bonf <- list()

for(file in names(gwas_list) ) {
  
  df <- gwas_list[[file]]
  name = names(gwas_list[file])
  
  bonf=1e-3/nrow(df) ## alpha = 10^-3 or alpha = 10^-8
  
  df_filter <- df %>%
    filter(pvalue <= bonf) %>%
    filter(!grepl('^unloc', Chromosome))
  
  gwas_filt_bonf[[name]] <- df_filter
  
}


##  Function to extract and combine required columns
extract_and_combine <- function(df) {
  selected_columns <- df[, c("Chr", 'Position', "pvalue", 'Chromosome')]
  return(selected_columns)
}

##  Apply the function to each data frame in the list
selected_data <- lapply(gwas_filt_bonf, extract_and_combine)

##  Remove data frames with 0 rows
selected_data_significant <- selected_data[sapply(selected_data, function(df) nrow(df) > 0)]


##  Add the data frame names as a new column
selected_data_sign_names <- Map(cbind, 
                                Name = names(selected_data_significant), 
                                selected_data_significant)

##  Combine the data frames into a single table
final_table <- do.call(rbind, selected_data_sign_names)


## Group for the summary
grouped_out <- final_table %>%
  group_by(Name, Chromosome) %>%
  summarise(Position_count = n_distinct(Position),
            Pops = paste(unique(Name), collapse = ","))

```


## Make GR object 
```{r }

granges_object_list <- GRangesList()
granges_object_red_list  <- GRangesList()

for(file in names(selected_data_sign_names)) {
  
  ## get names
  df <- selected_data_significant[[file]] 

  ## transform into GRanges object 
  SNPs_of_interest_ranges <- GRanges(seqnames = df$Chromosome,
                                     ranges = IRanges(start = df$Position,
                                                      end = df$Position),
                                     metadata = file,
                                     significant = df$pvalue)
## Merge the SNPs into regions, so that SNPs on the same chromosome and within 100 kb form a single block
  SNPs_of_interest_ranges_red <- GenomicRanges::reduce(SNPs_of_interest_ranges, min.gapwidth = 100000)
  
  
  ## Add to the list, also add names
  granges_object_list[[file]] <- SNPs_of_interest_ranges
  granges_object_red_list[[file]]   <- SNPs_of_interest_ranges_red
  
}


```


## Part 2: Gene overlapped with the outlier SNPs
```{r }
## Load GFF file
gff <- rtracklayer::import('/Users/rabu0002/Desktop/Comp_UU/Mariano_infor_REF_Genome_Dropoff-24DMUW5h4qGsmcXn/NBI Dropoff-c3C57sHTpWVaaNda/functional_annotation/Saalp_EIv1.0.release.functional_annotation.gff3')

##  Filter out entries where the seqnames start with "unloc"
gff <- gff[!grepl("^unloc", seqnames(gff)), ]


## Select genes and cds regions
gene <- gff[gff$type %in% c("gene","mRNA") ]
cds <- gff[gff$type == "CDS"]

## Find overlaps approach
findOverlaps_list <- list()
findOverlaps_list_data <- list()

for(file in names(granges_object_list) ){
  
  ## get names
  df <- granges_object_list[[file]] 
  
  ## Subset by overlaps
  findOverlaps_snp <- findOverlaps(gene,df,maxgap = 5000) ##  gene, cds, gff 

  subsOverlaps_snp<-subsetByOverlaps(gene,df, maxgap = 5000)

  ## Subset by overlaps
  subset_SNPs <- gene[findOverlaps_snp@from]
  subset_SNPs$regions <- df[findOverlaps_snp@to]
  subset_SNPs_data <- as.data.frame(subset_SNPs)
  
  ## Select columns
  ## subset_SNPs_data_sub <- 
  ##   subset_SNPs_data %>% 
  ##   dplyr::select('start', 'end', 'width',"Name" , 'biotype', 'regions.seqnames', 'regions.start', 'regions.width', 'regions.metadata')
  
  ## Save into file
  findOverlaps_list[[file]] <-  findOverlaps_snp
  findOverlaps_list_data[[file]] <- subset_SNPs_data
}

```



## Read the file names with gene symbols
```{r }
## Read the file names with genes
functional_code <- read.csv2("/Users/rabu0002/Desktop/Comp_UU/Mariano_infor_REF_Genome_Dropoff-24DMUW5h4qGsmcXn/NBI Dropoff-c3C57sHTpWVaaNda/functional_annotation/Saalp_EIv1.0.release.gff3.pep.fasta.functional_annotation.tsv", sep = '\t', header = F)
colnames(functional_code) <- c(functional_code[1,])
functional_code <- functional_code[-1,]

## Add names of the genes in human readable format
findOverlaps_list_data_gene <- list()

for(file in names(findOverlaps_list_data) ){
  
  ## get names
  df <- findOverlaps_list_data[[file]] 
  
  ##  Create a new column in dataframe1 based on matches in dataframe2
  df$Human_readable_gene <- functional_code$`## Human-Readable-Description`[match(df$Name,
                                                                functional_code$`## Gene`)]
  
  findOverlaps_list_data_gene[[file]] <- df
  
}

```


## STEP 3: Get a table for each trait and save 
```{r }
##  Remove data frames with 0 rows
findOverlaps_list_data_matches <- findOverlaps_list_data_gene[sapply(
  findOverlaps_list_data_gene, function(df) nrow(df) > 0)]


##  Add the data frame names as a new column
findOverlaps_list_data_matches_names <- 
  Map(cbind, 
      Lake_contrast = names(findOverlaps_list_data_matches), 
      findOverlaps_list_data_matches)


##  Combine the data frames into a single table
findOverlaps_table <- do.call(rbind, findOverlaps_list_data_matches_names)


## Group by trait and genes
findOverlaps_table_group <- 
  findOverlaps_table %>%
  group_by(Lake_contrast, regions.seqnames, start, end, Human_readable_gene) %>%
  summarise(n_SNP = n_distinct(regions.start)
            )

## Group by gene only and add contrast "."
findOverlaps_table_group_gene_list <- 
  findOverlaps_table %>%
  group_by(regions.seqnames, start, end,regions.start, Human_readable_gene) %>%
  summarise(n_SNP = n_distinct(regions.start),
            Lake_contrast = paste(unique(Lake_contrast), collapse = ",")
            )

```


## Step 4:Get a table for inversion regions and a shared region between Myvatn and Thingvallavatn
## ## ## Set inversion positions here +- 5000 bp 
```{r }
##  Define Positions of interest with unique labels for repetitive chromosomes
Positions <- list(
  ##  Contrast LB vs SB
  chr4_1 = c(75250000 - 5000, 76130000 + 5000),
  chr5_1 = c(22300000 - 5000, 22750000 + 5000),
  chr9_1 = c(61300000 - 5000, 62110000 + 5000),
  chr17_1 = c(32450000 - 5000, 33200000 + 5000),
  
  ##  Contrast Benthic vs Pelagic
  chr1_1 = c(16300000 - 5000, 18600000 + 5000),
  chr1_2 = c(19500000 - 5000, 22200000 + 5000),
  chr3_1 = c(33500000 - 5000, 35800000 + 5000),
  chr3_2 = c(37350000 - 5000, 40600000 + 5000),
  chr8_1 = c(29000000 - 5000, 29950000 + 5000),
  chr9_2 = c(38400000 - 5000, 40800000 + 5000),
  chr14_1 = c(6300000 - 5000, 6900000 + 5000),
  chr40_1 = c(16250000 - 5000, 17000000 + 5000),
  
  ##  Shared region between B vs P from Myvatn and Thigvallavatn
  chr34_1 = c(18340000 - 5000, 18450000 + 5000)
)

##  Convert Positions to a data frame for easier processing
inversion_regions <- do.call(rbind, lapply(names(Positions), function(region) {
  chr_base <- gsub("_.*", "", region)  ##  Extract base chromosome (e.g., 'chr1')
  suffix <- gsub(".*_", "", region)   ##  Extract suffix (e.g., '1')
  coords <- Positions[[region]]
  
  data.frame(
    CHR = as.numeric(gsub("chr", "", chr_base)),  ##  Remove 'chr' prefix for numeric conversion
    Min = coords[1],
    Max = coords[2],
    Label = paste0(
      as.numeric(gsub("chr", "", chr_base)), "_",
      round(coords[1] / 1e6, 3), "_",
      round(coords[2] / 1e6, 3),
      "_", suffix  ##  Add suffix for uniqueness
    )
  )
}))

inversion_regions$Contrast <- c(rep("Thi_DB_vs_LB", 4),rep("Thi_DB_LB_vs_PL_Pi", 9) )

```

## Select files of interes
```{r }
## I am using this table 
## findOverlaps_list_data_matches_names


##  Filter the final table based on 
files_of_interest <-  c("Thi_DB_vs_LB","Thi_DB_LB_vs_PL_Pi")

##  Initialize lists to store final data
final_data_clean <- list()
final_data_na <- list()

##  Loop over files of interest
for (file in files_of_interest) {
  df <- findOverlaps_list_data_matches_names[[file]]
  regions <- inversion_regions[inversion_regions$Contrast == file,]
  
  ##  Initialize list for filtered regions
  filtered_regions <- list()
  
  ##  Filter GWAS results based on inversion/shared regions
  for (i in 1:nrow(regions)) {
    region <- regions[i, ]
    
    ##  Filter rows based on region criteria
    subset <- df[
      df$seqnames == region$CHR &
        df$start >= region$Min &
        df$end <= region$Max, 
    ]
    
    if (nrow(subset) > 0) {
      ##  Add region label to subset
      subset$regions <- region$Label
      filtered_regions[[i]] <- subset
    }
  }
  
  ##  Combine filtered regions into one data frame
  findOverlaps_list_data_inversions <- do.call(rbind, filtered_regions)
  
  ##  Group by gene ID
  findOverlaps_list_data_inversions_gene <- 
    findOverlaps_list_data_inversions[findOverlaps_list_data_inversions$type == "gene", ] %>%
    group_by(seqnames, ID, start, end, regions, Lake_contrast) %>%
    summarise(
      n_snps = n_distinct(regions.start),
      ##  Full_gene_name_gff = Human_readable_gene
      .groups = "drop"
    ) %>%
    arrange(seqnames, start)
  
  ##  Join with BLAST results
  overlapped_gene_symbols <- 
    findOverlaps_list_data_inversions_gene %>%
    left_join(blast_results_all_chr_best_hit_geneID, by = c("ID" = "GeneID")) %>%
    dplyr::select(
      seqnames, Gene_Symbol, ID, start, end, evalue, identity, best_hit,
      n_snps, regions, Lake_contrast
    ) %>%
    dplyr::distinct(Gene_Symbol, .keep_all = TRUE) %>%  ##  Remove duplicates based on Gene_Symbol

    arrange(seqnames, start)
  
  ##  Filter clean data
  overlapped_gene_symbols_clean <- 
    overlapped_gene_symbols[!is.na(overlapped_gene_symbols$Gene_Symbol),]
  
  ##  Assign a readable column name
  colnames(overlapped_gene_symbols_clean)[8] <- "SwissProt|AssesionNumber|Protein_Organism"
  
  ##  Store results in lists
  final_data_clean[[file]] <- overlapped_gene_symbols_clean
  final_data_na[[file]] <- overlapped_gene_symbols
}

##  Combine all clean and NA data into single data frames
combined_clean <- do.call(rbind, final_data_clean)
combined_na <- do.call(rbind, final_data_na)


##  Write combined clean data to an Excel file
write.csv(combined_clean,"~/Desktop/Comp_UU/Manuscript/Sumbition/Supplementary_Material/Table_S6_S7.csv",  row.names = FALSE)

```

