#  Extract and Visualize Sequencing Depth Across Individuals

## Step 1: Extract average coverage from bamqc reports
```{r}
## !/bin/bash -l

## SBATCH -A naiss2023-5-221
## SBATCH -p core
## SBATCH -n 3
## SBATCH --array=1-283
## SBATCH -t 10:00:00
## SBATCH -J depth
## SBATCH -e depth_%A_%a.err
## SBATCH -o dept_%A_%a.out
## SBATCH --mail-type=all
## SBATCH --mail-user=khrystyna.kurta@imbim.uu.se


## STEP 2: Determine directory 
cd $BAM_DIR


## AVERAGE depth 
output_file=$BAM_DIR/"coverage_summary.txt"

SAMPLE=$(ls *_marked_dups_1_2.bam | sed -n ${SLURM_ARRAY_TASK_ID}p)

DIR=$BAM_DIR/${SAMPLE}'_bamqc'

coverage_line=$(grep "mean coverageData =" $DIR/genome_results.txt | awk '{print $1 " " $2 " " $3 " " $4}')
```


## Step 2: Visualize individual coverage distribution
```{r}
##  Load and clean coverage data
cov <- read.table('~/Desktop/Comp_UU/REF_SalAlp_UK/Coverage/Coverage_summary.txt')
cov$Sample_ID <- gsub('_marked_dups_1_2.bam_bamqc', '', cov$V1)
cov$Cov_stats <- as.numeric(gsub('X', '', cov$V5))
cov$Ref <- "Arctic_charr_ref_fSalAlp1"

##  Summary statistics
summary(cov)


##  Load libraries
library(ggplot2)
library(ggpubr)

##  Compute mean and SD
mean_cov <- mean(cov$Cov_stats)
sd_cov_upper <- mean_cov + sd(cov$Cov_stats)
sd_cov_lower <- mean_cov - sd(cov$Cov_stats)

##  Plot coverage per sample
cov_plot <- cov %>%
  ggplot(aes(x = reorder(V1, Cov_stats), y = Cov_stats)) +
  geom_bar(stat = "identity", fill = "steelblue", color = "darkblue") +
  labs(x = "Individual", y = "Mean sequencing depth") +
  theme_classic2(base_size = 14) +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    plot.title = element_text(hjust = 0.5, face = "bold"),
    axis.title = element_text(face = "bold")
  ) +
  geom_hline(yintercept = mean_cov, color = "red") +
  geom_hline(yintercept = sd_cov_upper, linetype = "dashed", color = "blue") +
  geom_hline(yintercept = sd_cov_lower, linetype = "dashed", color = "blue") +
  annotate("text", label = paste("mean = ", round(mean_cov, 2)), x = 22, y = 2.5) +
  annotate("text", label = paste("mean +1SD = ", round(sd_cov_upper, 2)), x = 28, y = 3.5) +
  annotate("text", label = paste("mean -1SD = ", round(sd_cov_lower, 2)), x = 28, y = 1.6)

##  Save plot
ggsave("./REF_SalAlp_UK/Coverage/Supplementary_Fig.2.pdf", cov_plot, width = 8, height = 5)
```


## Step 3: Estimate sequencing depth using mosdepthacross all samples
```{r}
## !/bin/bash -l

## SBATCH -A uppmax2025-2-114
## SBATCH -p core -n 10
## SBATCH --array=1-285
## SBATCH -t 02-00:00:00
## SBATCH -J dpth
## SBATCH -e depth_%A_%a.err
## SBATCH -o depth_%A_%a.out
## SBATCH --mail-type=all
## SBATCH --mail-user=khrystyna.kurta@slu.se


##  Load required tools
module load bioinfo-tools
module load mosdepth/0.3.3

##  Reference genome path
REFGENOME=./fSalAlp1.1/Data_Package_fSalAlp1_assembly_20231024/fSalAlp1.1.hap1.cur.20231016.fasta

##  Move to base directory
cd $BASEDIR

##  Get BAM file name by array index
BAM=$(cat all_bam.list | sed -n ${SLURM_ARRAY_TASK_ID}p)
NAME=${BAM/.bam/}

##  Run mosdepth to calculate depth in 500kb windows
mosdepth --threads 10 --by 500000 "$NAME" "$BAM"

```


## Step 4: Plot average sequencing depth across genome
```{r}
##  Load libraries
library(tidyverse)
library(data.table)

##  Set directory with mosdepth results
dir <- './REF_SalAlp_UK/DEPTH/MOSDEPTH500kb'

##  Read all .bed.gz files
files <- list.files(path = dir, pattern = "*.regions.bed.gz", full.names = TRUE)
stopifnot(length(files) > 0)

##  Read and clean first file
merged <- fread(files[1])[!grepl("unloc", V1)][, .(chrom = V1, start = V2, end = V3, depth1 = V4)]

##  Merge in remaining depth values
for (i in 2:length(files)) {
  dt <- fread(files[i])[!grepl("unloc", V1)][, .(V4)]
  merged <- cbind(merged, dt)
}

##  Rename columns
names_file <- basename(files)
colnames(merged) <- c("chrom", "start", "end", names_file)

##  Calculate average depth across samples per window
merged[, mean_depth := rowMeans(.SD, na.rm = TRUE), .SDcols = 4:ncol(merged)]


## Plot manhattan
##  Prepare data for CMplot
library(CMplot)

manh_data <- as.data.table(merged)
manh_data$chrom <- as.numeric(as.character(manh_data$chrom))
manh_data <- manh_data[order(manh_data$chrom),]
colnames(manh_data)[c(1,2,3,289)] <- c("CHR", "start", "end", "P")
manh_data$SNP <- paste(manh_data$CHR, manh_data$end, sep = "_")

cm_data <- manh_data[, c("SNP", "CHR", "end", "P")]
colnames(cm_data) <- c("SNP", "CHR", "BP", "P")

##  Create plot
CMplot(
  cm_data,
  plot.type = "m",
  LOG10 = FALSE,
  col = c("grey30", "steelblue"),
  ylab = "Sequencing Depth",
  chr.den.col = NULL,
  cex = 0.5,
  axis.cex = 1,
  axis.lwd = 1.5,
  mar = c(3, 6, 1, 1),
  lab.cex = 1,
  ylim = c(0, 18),
  file = "jpg",
  dpi = 300,
  file.output = TRUE,
  verbose = TRUE,
  width = 10,
  height = 4
)
```


