


## Step 1: Plot the GWC for Large Benthic vs Small Benthic morphs
```{r}

require(ggplot2)
require(egg)
require(tidyverse)
require(ggpubr)
require(dplyr)
require(grid)
library(scales)

dir_gwas <- '~/Desktop/Comp_UU/REF_SalAlp_UK/GWAS/Data'
plot_gwas <- '~/Desktop/Comp_UU/REF_SalAlp_UK/GWAS/Plots'

#Load data 
gwasResults <- read.table(file.path(dir_gwas,"Thi_DB_vs_LB.pvalue"), header = T)

bonf_1 = -log10(0.001/nrow(gwas_list[[1]]))
bonf_2= -log10(0.00000001/nrow(gwas_list[[1]]))

#Order contigs
orderContigs = c("1","2","3","4","5","6",
                 "7", 
                 "8","9","10","11","12" ,"13","14","15", "16",
                 "17" ,"18","19","20","21",
                 "22", 
                 "23","24","25", "26","27","28",
                 "29","30",
                 "31","32","33","34","35","36","37","38", "39","40","NA")

#Plot

  colnames(gwasResults)[2] <- 'BP'
  colnames(gwasResults)[7] <- 'P'
  colnames(gwasResults)[8] <- 'CHR'
  
  gwasResults <- gwasResults
  gwasResults$CHR <- factor(gwasResults$CHR, levels = orderContigs)
  
  don <- gwasResults %>% 
    
    # Compute chromosome size
    group_by(CHR) %>% 
    summarise(chr_len=max(BP)) %>% 
    
    # Calculate cumulative position of each chromosome
    mutate(tot=cumsum(as.numeric(chr_len))-chr_len) %>%
    dplyr::select(-chr_len) %>%
    
    # Add this info to the initial dataset
    left_join(gwasResults, ., by=c("CHR"="CHR")) %>%
    
    # Add a cumulative position of each SNP
    arrange(as.character(CHR), as.numeric(BP)) %>%
    mutate(BPcum = as.numeric(BP)+ as.numeric(tot) )
  
  # Prepare the X axis. Indeed we do not want to display the cumulative position of SNP in bp, but just show the chromosome name instead.
  
  don$CHR <- factor(don$CHR, levels = orderContigs)
  axisdf = don %>% group_by(CHR) %>% summarize(center=( max(BPcum) + min(BPcum) ) / 2 )
  
  
  # Ready to make the plot using ggplot2:
  don$CHR <- factor(don$CHR, levels = orderContigs) 
  
  #Threshold bonferroni allowing 10% error
  bonf = -log10(0.01/nrow(don))
  bonf_2= -log10(0.00000001/nrow(don))
  max_y = max(-log10(don$P)) + 6
  min_y = bonf-3
  
  # Add stars at specified chromosomes (5, 9, 17, 4)
  peak_points <- don %>%
  filter(CHR %in% c(1,3,9,14,40)) %>%
  group_by(CHR) %>%
  slice_max(order_by = -log10(P), n = 1)
  
  # Filter axis labels to display every third chromosome
third_labels <- axisdf %>% 
  mutate(CHR = ifelse(row_number() %% 2 == 0, CHR, "")) # Show label only if row index is divisible by 3

  
  #Plot 
  manh <- ggplot(don, aes(x = BPcum, y = -log10(P) ) ) +
    
    # Show all points
    geom_point( aes(color=factor(CHR)), alpha=0.8, size=1) +
    scale_color_manual(values = rep(c("grey", "black"), 9000 )) +
    
    # custom X axis:
    scale_x_continuous(label = third_labels$CHR, breaks= axisdf$center) +
      scale_y_continuous(expand = c(0, 0), limits = c(0, 32), 
                       labels = label_number(accuracy = 0.2)) +    # remove space between plot area and x axis
    
    # Custom the theme:
    theme_classic(20) +
    theme(
      plot.title = element_text(face = 'bold', size = 20),
      axis.text.x = element_text(angle = 90),
      axis.title = element_text(size = 18),
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    )+
    labs(y ='-log10(P-value)', x = 'Scaffold')+
    geom_hline(yintercept = bonf_1, linetype="dashed", color = "red", lwd=1.5 )+
    geom_hline(yintercept = bonf_2, linetype="dashed", color = "blue", lwd=1.5 )+
    annotate("text", x = 0.5, y = 32, label = "a", 
           size = 8, hjust = 0, vjust = 1, fontface = "bold") +
    
    # Add star annotations at the peak points for chromosomes 5, 9, 14, and 17
    geom_point(data = peak_points, aes(x = BPcum, y = -log10(P)), 
             shape = 8, color = "red", size = 8, stroke = 1.5)


# Save the plot
ggsave('~/Desktop/Comp_UU/REF_SalAlp_UK/GWAS/GWAS_plot_Thin_LB_vs_SB.png', manh, 
       width = 13, height = 4, dpi = 300)

```


## Step 2: Plot Zoomed Genome-Wide Contrast (GWC) and LD Patterns: LB vs SB Morphs
```{r}
library(ggplot2)
library(ggpubr)
library(scales)
library(dplyr)
library(data.table)
library(cowplot)

#Dir
dir = "/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/GL_MAF_GENO_INV/LD_FILES_FULL/FOLDER_RDS"
ld_dir="/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/GL_MAF_GENO_INV/LD_FILES_FULL"


# Load GWAS data
data <- fread(file.path(dir, "Thi_DB_vs_LB.pvalue"))
data[, Position := as.numeric(Position)]
data[, Position_Mb := as.numeric(Position) / 1e6]


# Configuration for each chromosome
config <- list(
  list(chr = 4, ld_file_name = "Thin_CharrLowPass_GATKMethod_MinMAF0.05_Chr4_invLB_SB_ld.ld",
       minn = 75250000, 
       maxx = 76130000, label = "b"),
  list(chr = 5, ld_file_name = "Thin_CharrLowPass_GATKMethod_MinMAF0.05_Chr5_invLB_SB_ld.ld",
       minn = 22300000, 
       maxx = 22750000, label = "c"),
    list(chr = 9, ld_file_name = "Thin_CharrLowPass_GATKMethod_MinMAF0.05_Chr9_invLB_SB_ld.ld",
       minn = 61300000, 
       maxx = 62110000, label = "d"),
  list(chr = 17, ld_file_name = "Thin_CharrLowPass_GATKMethod_MinMAF0.05_Chr17_invLB_SB_ld.ld", 
       minn = 32450000, 
       maxx = 33200000, label = "e")
)

make_plot <- function(chr, ld_file_name, minn1, maxx1, label) {
  file <- fread(file.path(ld_dir,  ld_file_name),
                select = 1:4)
  setnames(file, c("Pos1", "Pos2", "Distance", "R2_Pearson"))
  
  file[, R2_Pearson := as.numeric(R2_Pearson)]
  file <- file[!is.na(R2_Pearson)]
  
  file[, Chr := as.numeric(sub(":.*", "", Pos1))]
  file[, Pos1 := as.numeric(sub(".*:", "", Pos1))]
  file[, Pos2 := as.numeric(sub(".*:", "", Pos2))]
  file[, Pos1_Mb := round(as.numeric(sub(".*:", "", Pos1)) / 1e6, 3)]
  file[, Pos2_Mb := round(as.numeric(sub(".*:", "", Pos2)) / 1e6, 3)]


  axis_min <- min(file$Pos2) + 5e5
  axis_max <- max(file$Pos2) - 5e5

  
  #Get a top SNP
top_snp <- data[Chromosome == chr & Position >= minn1 & Position <= maxx1][which.min(pvalue)]
top_pos <- top_snp$Position # Convert back to bp to match LD file

ld_with_top <- file[Pos1 == top_pos | Pos2 == top_pos, .(OtherPos = ifelse(Pos1 == top_pos, Pos2, Pos1), R2_Pearson)]

ld_with_top[, Position := OtherPos]
ld_with_top[, OtherPos := NULL]  # optional: drop the old column

ld_with_top[, Position := as.numeric(Position)]

#Select region of interest
gwas_region <- data[Chromosome == chr & Position >  axis_min & Position < axis_max ]  

# Merge with LD R2 for SNPs within highlighted region
highlight_region <- merge(
  gwas_region,
  ld_with_top,
  by = "Position",
  all.x = TRUE
)

  
  gwas_plot <- ggplot(gwas_region) +
    geom_point(aes(x = Position / 1e6, y = -log10(pvalue)), size = 1, color = "black") +
    geom_point(data = highlight_region, aes(x = Position / 1e6, y = -log10(pvalue), color = R2_Pearson), size = 1.2) +
    scale_color_gradientn(colors = c("yellow2", "red", "darkblue"), na.value = "darkgrey", name = expression(R^2)) +
    annotate("text", x = axis_min / 1e6, y = 28, label = label, size = 6, hjust = 0, vjust = 1, fontface = "bold") +
    geom_hline(yintercept = -log10(0.001 / nrow(data)), linetype = "dashed", color = "red") +
    geom_hline(yintercept = -log10(1e-8 / nrow(data)), linetype = "dashed", color = "blue") +
    scale_y_continuous(labels = label_number(accuracy = 0.1)) +
    scale_x_continuous(limits = c(axis_min, axis_max) / 1e6, 
                       breaks = seq(axis_min, axis_max, by = 5e5) / 1e6,
                       labels = label_number(accuracy = 0.2)) +
    theme_classic(base_size = 16) +
    theme(legend.position = 'none', axis.title.x = element_blank(), 
          axis.text.x = element_blank()) +
    labs(y = "-log10(P-value)")

  ld_plot <- 
    file %>%
    ggplot(aes(x = Pos1_Mb, y = Pos2_Mb, fill = R2_Pearson)) +
    geom_tile() +
    scale_fill_gradientn(colors = c("yellow2", "red", "darkblue"), na.value = "grey", name = expression(R^2)) +
    scale_x_continuous(labels = label_number(accuracy = 0.2), 
                       limits = c(axis_min, axis_max) / 1e6,
                       breaks = seq(axis_min, axis_max, by = 5e5) / 1e6) +
    scale_y_continuous(labels = label_number(accuracy = 0.2), 
                       limits = c(axis_min, axis_max) / 1e6) +
    labs(x = paste("Scaffold", chr, "position (Mb)"), y = "Position") +
    theme_classic(16) +
    theme(legend.title = element_text(size = 9),
          legend.text = element_text(size = 9),
          legend.key.size = unit(0.4, "cm"),
          legend.position = c(0.95, 0.50))
  
  list(gwas = gwas_plot, ld = ld_plot)
}

# Process each chromosome and collect plots
plots <- lapply(config, function(cfg) do.call(make_plot, cfg))

# Arrange and save
all_plots <- ggarrange(
  plots[[1]]$gwas, plots[[2]]$gwas,
  plots[[1]]$ld,   plots[[2]]$ld,
  plots[[3]]$gwas, plots[[4]]$gwas,
  plots[[3]]$ld,   plots[[4]]$ld,
  ncol = 2, nrow = 4
)

ggsave(file.path(ld_dir, 'GWAS_LD_inv_Thin_LB_DB.png'), all_plots, width = 12, height = 9, dpi = 300)

```


