
## Step 1: Plot the GWC for Benthic ve Pelagic morphs
```{r}
require(ggplot2)
require(egg)
require(tidyverse)
require(ggpubr)
require(dplyr)
require(grid)
library(scales)

dir_gwas <- '~/Desktop/Comp_UU/REF_SalAlp_UK/GWAS/Data'
plot_gwas <- '~/Desktop/Comp_UU/REF_SalAlp_UK/GWAS/Plots'

#Load data 
gwasResults <- read.table(file.path(dir_gwas,"Thi_DB_LB_vs_PL_Pi.pvalue"), header = T)

bonf_1 = -log10(0.001/nrow(gwas_list[[1]]))
bonf_2= -log10(0.00000001/nrow(gwas_list[[1]]))

#Order chromosomes
orderContigs = c("1","2","3","4","5","6",
                 "7", 
                 "8","9","10","11","12" ,"13","14","15", "16",
                 "17" ,"18","19","20","21",
                 "22", 
                 "23","24","25", "26","27","28",
                 "29","30",
                 "31","32","33","34","35","36","37","38", "39","40","NA")

#Plot Benthic vs Pelagic

  colnames(gwasResults)[2] <- 'BP'
  colnames(gwasResults)[7] <- 'P'
  colnames(gwasResults)[8] <- 'CHR'
  
  gwasResults <- gwasResults
  gwasResults$CHR <- factor(gwasResults$CHR, levels = orderContigs)
  
  don <- gwasResults %>% 
    
    # Compute chromosome size
    group_by(CHR) %>% 
    summarise(chr_len=max(BP)) %>% 
    
    # Calculate cumulative position of each chromosome
    mutate(tot=cumsum(as.numeric(chr_len))-chr_len) %>%
    dplyr::select(-chr_len) %>%
    
    # Add this info to the initial dataset
    left_join(gwasResults, ., by=c("CHR"="CHR")) %>%
    
    # Add a cumulative position of each SNP
    arrange(as.character(CHR), as.numeric(BP)) %>%
    mutate(BPcum = as.numeric(BP)+ as.numeric(tot) )
  
  # Prepare the X axis. Indeed we do not want to display the cumulative position of SNP in bp, but just show the chromosome name instead.
  
  don$CHR <- factor(don$CHR, levels = orderContigs)
  axisdf = don %>% group_by(CHR) %>% summarize(center=( max(BPcum) + min(BPcum) ) / 2 )
  
  
  # Ready to make the plot using ggplot2:
  don$CHR <- factor(don$CHR, levels = orderContigs) 
  
  #Threshold bonferroni allowing 10% error
  bonf = -log10(0.01/nrow(don))
  bonf_2= -log10(0.00000001/nrow(don))
  max_y = max(-log10(don$P)) + 6
  min_y = bonf-3
  
  # Add stars at specified chromosomes (5, 9, 17, 4)
  peak_points <- don %>%
  filter(CHR %in% c(1,3,9,14,40)) %>%
  group_by(CHR) %>%
  slice_max(order_by = -log10(P), n = 1)
  
  # Filter axis labels to display every third chromosome
third_labels <- axisdf %>% 
  mutate(CHR = ifelse(row_number() %% 2 == 0, CHR, "")) # Show label only if row index is divisible by 3

  
  #Plot 
  manh <- ggplot(don, aes(x = BPcum, y = -log10(P) ) ) +
    
    # Show all points
    geom_point( aes(color=factor(CHR)), alpha=0.8, size=1) +
    scale_color_manual(values = rep(c("grey", "black"), 9000 )) +
    
    # custom X axis:
    scale_x_continuous(label = third_labels$CHR, breaks= axisdf$center) +
      scale_y_continuous(expand = c(0, 0), limits = c(0, 32), 
                       labels = label_number(accuracy = 0.2)) +    # remove space between plot area and x axis
    
    # Custom the theme:
    theme_classic(20) +
    theme(
      plot.title = element_text(face = 'bold', size = 20),
      axis.text.x = element_text(angle = 90),
      axis.title = element_text(size = 18),
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    )+
    labs(y ='-log10(P-value)', x = 'Scaffold')+
    geom_hline(yintercept = bonf_1, linetype="dashed", color = "red", lwd=1.5 )+
    geom_hline(yintercept = bonf_2, linetype="dashed", color = "blue", lwd=1.5 )+
    annotate("text", x = 0.5, y = 32, label = "a", 
           size = 8, hjust = 0, vjust = 1, fontface = "bold") +
    
    # Add star annotations at the peak points for chromosomes 5, 9, 14, and 17
    geom_point(data = peak_points, aes(x = BPcum, y = -log10(P)), 
             shape = 8, color = "red", size = 8, stroke = 1.5)


# Save the plot
ggsave('~/Desktop/Comp_UU/REF_SalAlp_UK/GWAS/GWAS_plot_Thin_B_vs_P.png', manh, 
       width = 13, height = 4, dpi = 300)

```


## Step 2: Plot Zoomed Genome-Wide Contrast (GWC) and LD Patterns: LB vs SB Morphs
```{r}

library(ggplot2)
library(ggpubr)
library(scales)
library(dplyr)
library(data.table)
library(ggrastr)
library(cowplot)


gwas_dir = "/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/GL_MAF_GENO_INV/LD_FILES_FULL/FOLDER_RDS"
ld_dir = "/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/GL_MAF_GENO_INV/LD_FILES_FULL"


# Read GWAS data with fread
data <- fread(file.path(gwas_dir, "Thi_DB_LB_vs_PL_Pi.pvalue"))

# Adjust the Position column by dividing it by 1e6
data[, Position := as.numeric(Position)]
data[, Position_Mb := as.numeric(Position) / 1e6]


make_plots <- function(chr, ld_file, region_list, plot_label) {
  file <- fread(file.path(ld_dir, ld_file), select = 1:4)
  setnames(file, c("Pos1", "Pos2", "Distance", "R2_Pearson"))
  file[, R2_Pearson := as.numeric(R2_Pearson)]
  file <- file[!is.na(R2_Pearson)]
  
  file[, Chr := as.numeric(sub(":.*", "", Pos1))]
  file[, Pos1 := as.numeric(sub(".*:", "", Pos1))]
  file[, Pos2 := as.numeric(sub(".*:", "", Pos2))]
  file[, Pos1_Mb := Pos1 / 1e6]
  file[, Pos2_Mb := Pos2 / 1e6]

#Axis
padding <- if (chr == 14) {0} else if (chr == 40) {6e5} else if (chr == 8) {5e5} else {4e5}

axis_min <- min(file$Pos2) + padding
axis_max <- max(file$Pos2) - padding
  
  # Use smaller tick size for smaller chromosomes
  breaks_by <- if (chr == 1 || chr == 3) {
  1e6
} else if (chr == 14) {
  1e5
} else {
  5e5
}
  
  #Accuracy of the axis x lable
  accuracy_by <- if (chr == 1 || chr == 3) {1} else if (chr == 14) {0.1} else {
  0.1}

  
  # Get SNPs in the first region for top SNP selection
  region_bounds <- region_list[[1]]
  top_snp <- data[Chromosome == chr & Position >= region_bounds[1] & Position <= region_bounds[2]][which.min(pvalue)]
  top_pos <- top_snp$Position
  
  ld_with_top <- file[Pos1 == top_pos | Pos2 == top_pos, .(OtherPos = ifelse(Pos1 == top_pos, Pos2, Pos1), R2_Pearson)]
  ld_with_top[, Position := as.numeric(OtherPos)]
  ld_with_top[, OtherPos := NULL]
  
  gwas_region <- data[Chromosome == chr & Position > axis_min & Position < axis_max]
  highlight_region <- merge(gwas_region, ld_with_top, by = "Position", all.x = TRUE)
  
  gwas_plot <- ggplot(gwas_region) +
    geom_point(aes(x = Position/1e6, y = -log10(pvalue)), size = 1, color = "black") +
    geom_point(data = highlight_region, aes(x = Position/ 1e6, y = -log10(pvalue), color = R2_Pearson), size = 1.2) +
    scale_color_gradientn(colors = c("yellow2", "red", "darkblue"), na.value = "darkgrey", name = expression(R^2)) +
    annotate("text", x = axis_min / 1e6, y = 28, label = plot_label, size = 6, hjust = 0, vjust = 1, fontface = "bold") +
    theme_classic(16) +
    labs(y = "-log10(P-value)") +
    theme(legend.position = "none",
          axis.title.x = element_blank(),
          axis.text.x = element_blank(),
          axis.title.y = element_text(size = 17)) +
    geom_hline(yintercept =  -log10(0.001 / nrow(data)), linetype = "dashed", color = "red") +
    geom_hline(yintercept = -log10(1e-8 / nrow(data)), linetype = "dashed", color = "blue") +
    scale_y_continuous(labels = label_number(accuracy = 0.1)) +
    scale_x_continuous(
      limits = c(axis_min, axis_max) / 1e6,
      breaks = seq(axis_min, axis_max, by = breaks_by) / 1e6,
      labels = label_number(accuracy = accuracy_by)
    )
  
  
  ld_plot <- ggplot(file, aes(x = Pos1_Mb, y = Pos2_Mb, fill = R2_Pearson)) +
    geom_tile_rast() +
    scale_fill_gradientn(colors = c("yellow2", "red", "darkblue"), na.value = "grey", name = expression(R^2)) +
    theme_classic(16) +
    labs(x = paste("Scaffold", chr, "position (Mb)"), y = "Position") +
    theme(legend.title = element_text(size = 9),
          legend.text = element_text(size = 9),
          legend.key.size = unit(0.4, "cm"),
          legend.position = c(0.95, 0.50),
          axis.title.y = element_text(size = 17)) +
    scale_x_continuous(limits = c(axis_min, axis_max) / 1e6,
                       breaks = seq(axis_min, axis_max, by = breaks_by) / 1e6,
                       labels = label_number(accuracy = accuracy_by)) +
    scale_y_continuous(limits = c(axis_min, axis_max) / 1e6,
                       labels = label_number(accuracy = 0.1))
  
list(gwas = gwas_plot, ld = ld_plot)
}

# Configuration for each chromosome
config_list <- list(
  list(chr = 1, ld_file = "Thin_CharrLowPass_GATKMethod_MinMAF0.05_Chr1_invB_P_ld_0.3.ld",
       regions = list(c(16300000, 18600000), c(19500000, 22200000)), label = "b"),
  list(chr = 3, ld_file = "Thin_CharrLowPass_GATKMethod_MinMAF0.05_Chr3_invB_P_ld.ld",
       regions = list(c(33500000, 35800000), c(37350000, 40600000)), label = "c"),
  list(chr = 8, ld_file = "Thin_CharrLowPass_GATKMethod_MinMAF0.05_Chr8_invB_P_ld.ld",
       regions = list(c(29050000, 29830000)), label = "d"),
  list(chr = 9, ld_file = "Thin_CharrLowPass_GATKMethod_MinMAF0.05_Chr9_invB_P_ld.ld",
       regions = list(c(38400000, 40800000)), label = "e"),
  list(chr = 14, ld_file = "Thin_CharrLowPass_GATKMethod_MinMAF0.05_Chr14_invB_P_ld.ld",
       regions = list(c(6330000, 6870000)), label = "f"),
  list(chr = 40, ld_file = "Thin_CharrLowPass_GATKMethod_MinMAF0.05_Chr40_invB_P_ld.ld",
       regions = list(c(16250000, 17010000)), label = "g")
)


plots <- lapply(config_list, function(cfg) {
  make_plots(cfg$chr, cfg$ld_file, cfg$regions, cfg$label)
})


final_plot <- plot_grid(
  plots[[1]]$gwas, plots[[2]]$gwas,
  plots[[1]]$ld,   plots[[2]]$ld,
  plots[[3]]$gwas, plots[[4]]$gwas,
  plots[[3]]$ld,   plots[[4]]$ld,
  plots[[5]]$gwas, plots[[6]]$gwas,
  plots[[5]]$ld,   plots[[6]]$ld,
  ncol = 2, nrow = 6, align = "v"
)

ggsave(file.path(ld_dir, 'GWAS_LD_inv_Thin_BvsP.png'), final_plot, width = 12, height = 11, dpi = 300)


```

