
# PCA plot of genotype distributions among Arctic charr benthic and pelagic morphs in Lake Thingvallavatn
```{r setup, include=FALSE}

require(tidyverse)
library(psych)
library(tidyr)
library(dplyr)
library(ggplot2)
library(RColorBrewer)
library(egg)
require(ggforce)

#Read bam list with lakes file 
bams <- read.csv("~/Desktop/Comp_UU/REF_SalAlp_UK/BAM_list/all_bam_pca_lakes.list", header = T)

Thin_bam_list <- read.csv("~/Desktop/Comp_UU/REF_SalAlp_UK/Admixture/ADM_INV_THIN/Thin.list", header = F)

#Merge the information above 
bamList <- right_join(bams, Thin_bam_list, by = c("Bam" ="V1"))


#Add pcangsd matrix
dir_pca <- '~/Desktop/Comp_UU/REF_SalAlp_UK/Admixture/THIN_INV_PCA'
cov_mat <- list.files(dir_pca, pattern = "cov$")

#Add homo/hetero list of Myvatn morphs
dir_geno_list <- "~/Desktop/Comp_UU/REF_SalAlp_UK/Inversions/Bam_homo_list"


#Configure file
# Define inversion config: file name, chromosome, start, end, pretty title
inversion_config <- list(

  list(
    file = "Thin_RefUK_MAF0.05_MapQ30_chunk_inv_1chr_16.3_18.8.cov",
    chr = 1,
    start = 16300000,
    end = 18600000,
    title = "1: 16.3-18.8 Mb",
    ind_list_homoB = "ind_list_homB_chr_1_part1.txt",
    ind_list_homoP = "ind_list_homP_chr_1_part1.txt",
    ind_list_het = "ind_list_homHet_chr_1_part1.txt"
  ),
  list(
    file = "Thin_RefUK_MAF0.05_MapQ30_chunk_inv_1chr_19.5_22.2.cov",
    chr = 1,
    start = 19500000,
    end = 22200000,
    title = "1: 19.5-22.2 Mb",
    ind_list_homoB = "ind_list_homB_chr_1_part2.txt",
    ind_list_homoP = "ind_list_homP_chr_1_part2.txt",
    ind_list_het = "ind_list_homHet_chr_1_part2.txt"
  ),
  list(
    file = "Thin_RefUK_MAF0.05_MapQ30_chunk_inv_3chr_33_35.8.cov",
    chr = 3,
    start = 33500000,
    end = 35800000,
    title = "3: 33.0-35.8 Mb",
    ind_list_homoB = "ind_list_homB_chr_3_part1.txt",
    ind_list_homoP = "ind_list_homP_chr_3_part1.txt",
    ind_list_het = "ind_list_homHet_chr_3_part1.txt"
  ),
  list(
    file = "Thin_RefUK_MAF0.05_MapQ30_chunk_inv_3chr_37.35_40.6.cov",
    chr = 3,
    start = 37350000,
    end = 40600000,
    title = "3: 37.3-40.6 Mb",
    ind_list_homoB = "ind_list_homB_chr_1_part2.txt",
    ind_list_homoP = "ind_list_homP_chr_1_part2.txt",
    ind_list_het = "ind_list_homHet_chr_1_part2.txt"
  ),
  list(
    file = "Thin_RefUK_MAF0.05_MapQ30_chunk_inv_8chr_29_29.95.cov",
    chr = 8,
    start = 29050000, 
    end = 29830000,
    title = "8: 29.05-29.83 Mb",
    ind_list_homoB = "ind_list_homB_chr_8_part_1.txt",
    ind_list_homoP = "ind_list_homP_chr_8_part_1.txt",
    ind_list_het = "ind_list_homHet_chr_8_part_1.txt"
  ),
  list(
    file = "Thin_RefUK_MAF0.05_MapQ30_chunk_inv_9chr_38.4_40.8.cov",
    chr = 9,
    start = 38400000,
    end = 40800000,
    title = "9: 38.40-40.80 Mb",
    ind_list_homoB = "ind_list_homB_chr_9_part_1.txt",
    ind_list_homoP = "ind_list_homP_chr_9_part_1.txt",
    ind_list_het = "ind_list_homHet_chr_9_part_1.txt"
  ),
  list(
    file = "Thin_RefUK_MAF0.05_MapQ30_chunk_inv_14chr_6.3_6.9.cov",
    chr = 14,
    start = 6330000, 
    end =  6870000,
    title = "14: 6.33-6.87 Mb",
    ind_list_homoB = "ind_list_homB_chr_14_part_1.txt",
    ind_list_homoP = "ind_list_homP_chr_14_part_1.txt",
    ind_list_het = "ind_list_homHet_chr_14_part_1.txt"
  ),
  list(
    file = "Thin_RefUK_MAF0.05_MapQ30_chunk_inv_40chr_16.2_17.cov",
    chr = 40,
    start = 16250000, 
    end = 17010000,
    title = "40: 16.25-17.01 Mb",
    ind_list_homoB = "ind_list_homB_chr_40_part_1.txt",
    ind_list_homoP = "ind_list_homP_chr_40_part_1.txt",
    ind_list_het = "ind_list_homHet_chr_40_part_1.txt"
  )

)

#Run extimates
cov_plots <- list()

# Loop over them
for (config in inversion_config) {
  
  
  # You can use cfg$file, cfg$chr, cfg$start, cfg$end in your plotting or analysis.
  data <- read.table(file.path(dir_pca, config$file), header = F)
  name = config$title
  name = basename(name)
  
    # Genotype class assignment
  homB <- read.table(file.path(dir_geno_list, config$ind_list_homoB), header = F)
  homB$Genotype <- "Homozygous major"
  homP <- read.table(file.path(dir_geno_list, config$ind_list_homoP), header = F)
  homP$Genotype <-"Homozygous minor" 
  het  <- read.table(file.path(dir_geno_list, config$ind_list_het), header = F)
  het$Genotype <-"Heterozygous"
  shape_info <- rbind(homB, homP, het)

#Run this uniquly
eigen.data <- eigen(data)

#We can then extract the eigenvectors from the pca object and format them into a dataframe for plotting, e.g. using ggplot().
eigenvectors <- as.data.frame(eigen.data$vectors)
eigenvalues <-  eigen.data$values

eigenvectors$Sample <- bamList$Bam
eigenvectors$Morph_short <- bamList$Morph_short
eigenvectors$Lake <-  bamList$Lake
eigenvectors$Group <- bamList$Morph_group
eigenvectors <- left_join(eigenvectors, shape_info, by = c("Sample" = "V1"))


#Get vars
pca.eigenval.sum = sum(eigen.data$values)
varPC1 <- (eigen.data$values[1]/pca.eigenval.sum)*100 #Variance explained by PC1
varPC2 <- (eigen.data$values[2]/pca.eigenval.sum)*100 #Variance explained by PC2
varPC3 <- (eigen.data$values[3]/pca.eigenval.sum)*100 #Variance explained by PC3
varPC4 <- (eigen.data$values[4]/pca.eigenval.sum)*100 #Variance explained by PC4

plt <-ggplot(data = na.omit(eigenvectors), 
              aes(x = V1, y = V2, shape = Group, color = Genotype)) +
  geom_point(alpha = 0.9, size = 3) +
  xlab(paste0("PC1: ", round(varPC1,1),"% variance")) +
  ylab(paste0("PC2: ", round(varPC2,1),"% variance")) +
  theme_bw() +
  ylim(-0.6,0.6)+
  xlim(-0.25,0.25)+
  theme(text=element_text(size=15)) +
  labs(color = 'Genotype', shape = "Morph group", title = paste(name))+
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.position = "bottom",
        plot.title = element_text(face = "bold")) +
  scale_color_manual(values = c("Homozygous minor" = "#67a9cf", 
                                "Homozygous major" =  "brown",
                                "Heterozygous" = "yellow3"))+
  geom_vline(xintercept = 0, linetype = "dotted")+
  geom_hline(yintercept = 0, linetype = "dotted")
  
  cov_plots[[name]] <- plt
  
}

final_plot <- do.call(ggarrange, c(
  cov_plots,
  list(
    ncol = 2,
    nrow = 4,
    labels = letters[1:length(cov_plots)],
    common.legend = TRUE,
    legend = "bottom"
  )
))

#Save
ggsave('/Users/rabu0002/Desktop/Comp_UU/REF_SalAlp_UK/Admixture/Figure.S10.pdf', final_plot, width = 11, height = 11, dpi = 300)
```


# PCA plot of genotype distributions among Arctic charr large and small benthivorous morphs in Lake Thingvallavatn
```{r}
#Add pcangsd matrix
dir_pca <- '~/Desktop/Comp_UU/REF_SalAlp_UK/Admixture/THIN_INV_PCA'
cov_mat <- list.files(dir_pca, pattern = "cov$")

dir_geno_list <- '~/Desktop/Comp_UU/REF_SalAlp_UK/Inversions/Bam_homo_list_benthic'

#Configure file
# Define inversion config: file name, chromosome, start, end, pretty title
inversion_config <- list(

  list(
    file = "Thin_Charr_RefUK_MAF0.05_chunk_inv_reg_chr4_LBSB.cov",
    chr = 4,
    title = "4: 75.25-76.13 Mb",
    ind_list_homoSB = "list_homSB.chr.4.txt",
    ind_list_homoLB = "list_homLB.chr.4.txt",
    ind_list_het = "list_het.chr.4.txt"
  ),
  list(
    file = "Thin_Charr_RefUK_MAF0.05_chunk_inv_reg_chr5_LBSB.cov",
    chr = 5,
    title = "5: 22.30-22.75 Mb",
    ind_list_homoSB = "list_homSB.chr.5.txt",
    ind_list_homoLB = "list_homLB.chr.5.txt",
    ind_list_het = "list_het.chr.5.txt"
  ),
  list(
    file = "Thin_Charr_RefUK_MAF0.05_chunk_inv_reg_chr9_LBSB.cov",
    chr = 9,
    title = "9: 61.30-62.11 Mb",
    ind_list_homoSB = "list_homSB.chr.9.txt",
    ind_list_homoLB = "list_homLB.chr.9.txt",
    ind_list_het = "list_het.chr.9.txt"
  ),
  list(
    file = "Thin_Charr_RefUK_MAF0.05_chunk_inv_reg_chr17_LBSB.cov",
    chr = 17,
    title = "17: 32.45-33.20 Mb",
    ind_list_homoSB = "list_homSB.chr.9.txt",
    ind_list_homoLB = "list_homLB.chr.9.txt",
    ind_list_het = "list_het.chr.9.txt"
  )
  )

#Bam list
bamL <- bamList[bamList$Morph_group == "Benthic",]


#Run extimates
cov_plots <- list()

# Loop over them
for (config in inversion_config) {
  
  
  # You can use cfg$file, cfg$chr, cfg$start, cfg$end in your plotting or analysis.
  data <- read.table(file.path(dir_pca, config$file), header = F)
  name = config$title
  name = basename(name)
  
    # Genotype class assignment
  homB <- read.table(file.path(dir_geno_list, config$ind_list_homoSB), header = F)
  homB$Genotype <- "Homozygous major"
  homP <- read.table(file.path(dir_geno_list, config$ind_list_homoLB), header = F)
  homP$Genotype <-"Homozygous minor" 
  het  <- read.table(file.path(dir_geno_list, config$ind_list_het), header = F)
  het$Genotype <-"Heterozygous"
  shape_info <- rbind(homB, homP, het)

#Run this uniquly
eigen.data <- eigen(data)

#We can then extract the eigenvectors from the pca object and format them into a dataframe for plotting, e.g. using ggplot().
eigenvectors <- as.data.frame(eigen.data$vectors)
eigenvalues <-  eigen.data$values

eigenvectors$Sample <- bamL$Bam
eigenvectors$Morph_short <- bamL$Morph_short
eigenvectors$Lake <-  bamL$Lake
eigenvectors$Group <- bamL$Morph_group
eigenvectors <- left_join(eigenvectors, shape_info, by = c("Sample" = "V1"))


#Get vars
pca.eigenval.sum = sum(eigen.data$values)
varPC1 <- (eigen.data$values[1]/pca.eigenval.sum)*100 #Variance explained by PC1
varPC2 <- (eigen.data$values[2]/pca.eigenval.sum)*100 #Variance explained by PC2
varPC3 <- (eigen.data$values[3]/pca.eigenval.sum)*100 #Variance explained by PC3
varPC4 <- (eigen.data$values[4]/pca.eigenval.sum)*100 #Variance explained by PC4

plt <-ggplot(data = na.omit(eigenvectors), 
              aes(x = V1, y = V2, shape = Morph_short, color = Genotype)) +
  geom_point(alpha = 0.9, size = 3) +
  xlab(paste0("PC1: ", round(varPC1,1),"% variance")) +
  ylab(paste0("PC2: ", round(varPC2,1),"% variance")) +
  theme_bw() +
  ylim(-0.6,0.6)+
  xlim(-0.25,0.25)+
  theme(text=element_text(size=15)) +
  labs(color = 'Genotype', shape = "Morph group", title = paste(name))+
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.position = "bottom",
        plot.title = element_text(face = "bold")) +
  scale_color_manual(values = c("Homozygous minor" = "#67a9cf", 
                                "Homozygous major" =  "brown",
                                "Heterozygous" = "yellow3"))+
  geom_vline(xintercept = 0, linetype = "dotted")+
  geom_hline(yintercept = 0, linetype = "dotted")
  
  cov_plots[[name]] <- plt
  
}

final_plot <- do.call(ggarrange, c(
  cov_plots,
  list(
    ncol = 2,
    nrow = 2,
    labels = letters[1:length(cov_plots)],
    common.legend = TRUE,
    legend = "bottom"
  )
))

#Save
ggsave('/Users/rabu0002/Desktop/Comp_UU/REF_SalAlp_UK/Admixture/Figure.S8.pdf', final_plot, width = 11, height = 7, dpi = 300)
```





