
# Predicted genotypes based on genotype likelihoods from diagnostic markers at four putative inversions distinguishing large and small benthivorous morphs in Thingvallavatn
## Step 1: Load geno files
```{r}
require(tidyverse)
require(ggplot2)
require(ggrastr)
require(raster)

#Set up directory
dir = '~/Desktop/Comp_UU/REF_SalAlp_UK/Inversions/Geno'

#Load sample names 
bams <- read.csv("~/Desktop/Comp_UU/REF_SalAlp_UK/BAM_list/all_bam_pca_lakes.list")

#Load data 
files <- list.files(dir)

# Filter files that ends with LBSB.region.geno.gz
desired_files <- grep("LBSB.region.geno.gz$", files, value = TRUE)

#Check
desired_files

# Read the desired files 
geno_list <- list()

for(file in desired_files) {
  file_content <- read.table(file.path(dir, file), header = F)
  
  #Add colnames
  colnames(file_content)[3:287] <- bams$SampleID
  colnames(file_content)[1] <- 'Chromosome'
  colnames(file_content)[2] <- 'Position'
  
  file_content$SNP_ID <- paste(file_content$Chromosome, file_content$Position, sep = "_")
  
  # Save
  geno_list[[file]] <- file_content
  
}

# Combine all data frames into one
geno_all <- bind_rows(geno_list)

#Remove two Myv-12 and MyvK-20
remove_list <- c("Myv-12", "MyvK-20")
geno_all <- geno_all[, !colnames(geno_all) %in% remove_list ]


```



## Step 2: Load GWAS file for Benthic groups to select diagnistic SNPs
```{r }
dir_gwas <- '~/Desktop/Comp_UU/REF_SalAlp_UK/GWAS/Data'

gwas <- read.table(file.path(dir_gwas, 'Thi_DB_vs_LB.pvalue'), header = T)

#Add filter parameters
bonf = 0.00000001/nrow(gwas) #alpha = 10-8

#bonf = 0.001/nrow(gwas_Thi_DB_LB_vs_PL_Pi) #or alpha 0.001

chr_inv_list <- c('4','5','9','17')

#Filter
gwas_bonf <- gwas[gwas$pvalue < bonf,]

#Filter based on chr of interest
gwas_bonf_inv <- gwas_bonf[ gwas_bonf$Chromosome %in% chr_inv_list, ]


```


## Step 3: Filter set up: all Positions highlighting inversions
```{r }

#Add bonf filtered data
gwas_data <- gwas_bonf_inv

# Define Positions of interest
Positions <- list(
  chr4 = c(75250000,76130000),
  chr5 = c(22300000,22750000),
  chr9 = c(61300000,62110000),
  chr17 = c(32450000,33200000)
)

# Filter the GWAS data based on Positions
filtered_gwas_data <- gwas_data[
  (gwas_data$Chromosome == 4 & gwas_data$Position >= Positions$chr4[1] & gwas_data$Position <= Positions$chr4[2]) |
  (gwas_data$Chromosome == 5 & gwas_data$Position >= Positions$chr5[1] & gwas_data$Position <= Positions$chr5[2]) |
  (gwas_data$Chromosome == 9 & gwas_data$Position >= Positions$chr9[1] & gwas_data$Position <= Positions$chr9[2]) |
  (gwas_data$Chromosome == 17 & gwas_data$Position >= Positions$chr17[1] & gwas_data$Position <= Positions$chr17[2]) ,
]

#Add SNP Id column
filtered_gwas_data$SNP_ID <- paste(filtered_gwas_data$Chromosome, filtered_gwas_data$Position, sep = "_")


```

#Structure the data
```{r }
#Filtered
chr_data = filtered_gwas_data

#Make a longer matrix
geno_sign_position <- geno_all[geno_all$SNP_ID %in% chr_data$SNP_ID, ]

# Convert to longer format:
geno_sign_position %>%
  pivot_longer(cols=3:285,names_to="individuals",values_to="genotypes") -> table_geno_sign_position  # Make sure your number of columns is correct!! I have 91 individuals, so I want to pivot columns 3 to 93, since first two columns are the CHROM and POS.

table_geno_sign_position$Position <- as.character(table_geno_sign_position$Position)


#MAKE A MATRIX Each row - one individual
library(tidyverse)
geno_sign_position_heatmap <- 
  pivot_wider(table_geno_sign_position[, -c(1,2)],
              names_from = SNP_ID,
              values_from = genotypes)


#Add sample information 
geno_sign_position_heat_id <- 
  geno_sign_position_heatmap %>% 
  left_join(bams, by = c('individuals' = 'SampleID'))
geno_sign_position_heat_id <- geno_sign_position_heat_id[! (geno_sign_position_heat_id$individuals) %in% c( 'Myv-12', 'MyvK-20'), ]


```

#Heatmap preps
```{r }
#Order data based on levels 

level.order = c('ThiSB', 'ThiLB', 'ThiP', 'ThiPL',"MyvSB", 'MyvLG', 'SirD', 'SirN', 'VanD', 'VanN')

#Order data
geno_sign_position_heat_id_order <- geno_sign_position_heat_id[order(match(geno_sign_position_heat_id$Morph_lake, level.order)),]

#Choose reference samples (in our case it is  benthic morph SB)
geno_sign_position_heat_Benthic_Thin <- 
  geno_sign_position_heat_id[geno_sign_position_heat_id$Morph_lake == 'ThiSB', ]

#Replace NA in each column 
geno_sign_position_heat_Benthic_Thin[geno_sign_position_heat_Benthic_Thin == -1] <- NA

```


### Select reference
```{r }

#Add row sum row at the end - a reference column
geno_sign_position_heat_Benthic_Thin <- 
  geno_sign_position_heat_Benthic_Thin %>%
  bind_rows(summarise_all(., ~if(is.numeric(.)) sum(.,na.rm = T) else "RefSumGeno"))

#Add number of NAs in each column
geno_sign_position_heat_Benthic_Thin_sumNA <- 
  geno_sign_position_heat_Benthic_Thin %>%
  bind_rows(summarise_all(., ~if(is.numeric(.)) sum(is.na(.)) else "RefSumNA"))


#select only two row sums
RefSumGeno <- geno_sign_position_heat_Benthic_Thin_sumNA[(nrow(geno_sign_position_heat_Benthic_Thin_sumNA)-1):nrow(geno_sign_position_heat_Benthic_Thin_sumNA),] 

#invert the matrix
RefSumGeno <- data.frame(t(RefSumGeno))          


#how many nonNA colums = nInd -  number of NAs
RefSumGeno$RefSumNoNA <- (nrow(geno_sign_position_heat_Benthic_Thin_sumNA)-2) - as.numeric(RefSumGeno$X2) 


#Account for NA values 
RefSumGeno$RefSumGeno_with_geno <- round(as.numeric(RefSumGeno$X1)/(RefSumGeno$RefSumNoNA*2),1)

#Select all above 0.5
RefSumGenoSNP_above05 <- RefSumGeno[RefSumGeno$RefSumGeno_with_geno >= 0.5, ]
RefSumGenoSNP_above05_list <- data.frame(rownames(na.omit(RefSumGenoSNP_above05)))
colnames(RefSumGenoSNP_above05_list) <- 'SNP_list'


```

### This is modified chunk
```{r }
#Select data based on SNP list
geno_sign_position_heat_id_SelectRefSNP <- geno_sign_position_heat_id 

#Replace NA
geno_sign_position_heat_id_SelectRefSNP[geno_sign_position_heat_id_SelectRefSNP == -1] <- NA

# Check if column names are in the SNP list
cols_to_subtract <- colnames(geno_sign_position_heat_id_SelectRefSNP) %in% RefSumGenoSNP_above05_list$SNP_list

# Subtract genotypes from 2 where necessary
geno_sign_position_heat_id_SelectRefSNP[, cols_to_subtract] <- 2 - geno_sign_position_heat_id_SelectRefSNP[, cols_to_subtract]

geno_sign_position_heat_id_SelectRefSNP_AltGeno <- geno_sign_position_heat_id_SelectRefSNP


#Replace NA back to -1
geno_sign_position_heat_id_SelectRefSNP_AltGeno[is.na(geno_sign_position_heat_id_SelectRefSNP_AltGeno)] <- -1

#Prepare to plot Heat map
unique(geno_sign_position_heat_id_SelectRefSNP_AltGeno$Morph_lake)

#Order data based on levels 
level.order = c('ThiSB', 'ThiLB', 'ThiP', 'ThiPL',"MyvSB", 'MyvLG', 'SirD', 'SirN', 'VanD', 'VanN')

geno_sign_position_heat_id_SelectRefSNP_id_order <- geno_sign_position_heat_id_SelectRefSNP_AltGeno[order(match(geno_sign_position_heat_id_SelectRefSNP_AltGeno$Morph_lake, level.order)),]

#Prepare matrix
n_snps = dim(geno_sign_position_heat_id_SelectRefSNP_AltGeno)[2]-7
datamatrix <- 
  as.matrix(
    geno_sign_position_heat_id_SelectRefSNP_id_order[,2:n_snps])

#Add rownames                                             
row.names(datamatrix) <- unique(geno_sign_position_heat_id_order$individuals)

#Add row annotations
sampleInfo <- 
  geno_sign_position_heat_id_SelectRefSNP_id_order[,c((n_snps+1):ncol(geno_sign_position_heat_id_SelectRefSNP_id_order))]

sampleInfo <- data.frame(sampleInfo)
row.names(sampleInfo) <- row.names(datamatrix)

#Check the number of samples
#dim(datamatrix)
#dim(sampleInfo)


#Plot heatmap for all inversions

# Load necessary libraries
library(ComplexHeatmap)
library(circlize)

# Assuming your data matrix and sampleInfo are loaded
# datamatrix <- ... # Matrix with genotype data (individuals as rows, SNPs as columns)
# sampleInfo <- ... # Data frame with columns for 'individual', 'Lake', 'Morph'

haplotype_ranges <- list(
  "4: 75.2-76.1" = c("4_75250000", "4_76130000"),
  "5: 22.3-22.7" = c("5_22300000", "5_22750000"),
  "9: 61.3-62.1" = c("9_61300000", "9_62110000"),
  "17: 32.4-33.2" = c("17_32450000", "17_33200000")
)


# Initialize the column split vector
column_split <- character(ncol(datamatrix))

# Loop through each column to assign haplotype labels
for (i in seq_along(colnames(datamatrix))) {
  col_name <- colnames(datamatrix)[i]
  
  if (grepl("^4_", col_name)) {
    pos <- as.numeric(sub("4_", "", col_name))
    if (pos >= 75250000 && pos <= 76130000) {
      column_split[i] <- "4: 75.2-76.1"
    }
  } else if (grepl("^5_", col_name)) {
    pos <- as.numeric(sub("5_", "", col_name))
    if (pos >= 22300000 && pos <= 22750000) {
      column_split[i] <- "5: 22.3-22.7"
    }
  } else if (grepl("^9_", col_name)) {
    pos <- as.numeric(sub("9_", "", col_name))
    if (pos >= 61300000 && pos <= 62110000) {
      column_split[i] <- "9: 61.3-62.1"
    }
  } else if (grepl("^17_", col_name)) {
    pos <- as.numeric(sub("17_", "", col_name))
    if (pos >= 32450000 && pos <= 33200000) {
      column_split[i] <- "17: 32.4-33.2"
    }
  }
}

# Convert column_split to a factor
column_split <- factor(column_split,levels = c(  "4: 75.2-76.1",
  "5: 22.3-22.7",
  "9: 61.3-62.1",
  "17: 32.4-33.2"
)
 )

# Print lengths to debug
print(paste("Length of column_split:", length(column_split)))
print(paste("Number of columns in datamatrix:", ncol(datamatrix)))


# Prepare row split by 'Lake' (Factor levels should be set to maintain consistent order)
row_split <- factor(sampleInfo$Lake, 
                    levels = c("Thingvallavatn", "Mývatn", "Sirdalsvatnet", "Vangsvatnet"))

# Set up colors
color_list <- list(Lake = c(
  'Thingvallavatn' = 'forestgreen',
  'Mývatn' = "#ef8a62", 
  'Sirdalsvatnet' = "darkred",
  'Vangsvatnet' = 'gold'),
  Morph =  c( 
    "DB" = "blue",
    "LP" = "gold",
    'LB' = 'darkslateblue',
    'SB' = 'deepskyblue3',
    'Pi' = 'coral3',
    'PL' = 'chartreuse3',
    "LG" = "darkred"))


# Create row annotations for 'Lake' and 'Morph'
annot <- rowAnnotation(
    Morph = sampleInfo$Morph_short,
    col = color_list,
    show_annotation_name = FALSE,
    annotation_legend_param = list(
        Morph = list(title = "Morph", direction = "horizontal", 
                     title_gp = gpar(fontsize = 18),  # Increase title fontsize
                     labels_gp = gpar(fontsize = 18),
                     legend_width = unit(6, "cm"),
                     legend_height = unit(1, "cm")
                     
                     )  # Increase label fontsize
    ),
    gp = gpar(fontsize = 22)
     
)

# Customize heatmap legend to be displayed horizontally
heatmap_legend_param <- list(
    title = "Genotype",
    title_gp = gpar(fontsize = 16),  # Increase legend title size
    at = c(-1, 0, 1, 2),
    labels = c("NA", "0", "1", "2"),
    labels_gp = gpar(fontsize = 16),  # Increase legend label size
    heatmap_legend_param = list(legend_direction  = "horizontal", 
                                legend_width = unit(5, "cm"),
                                title_position = "lefttop",
                                grid_height = unit(1, "cm"),
                                grid_width = unit(1, "cm")
                                ) 
)


# Plot a heatmap using ComplexHeatmap
outdir="/Users/rabu0002/Desktop/Comp_UU/Manuscript/Sumbition/Supplementary_Figures"
pdf(file.path(outdir, paste('Supplementary_Fig.S7.pdf', sep = ".")), 
    width = 14, height = 13)
Heatmap(
  datamatrix,
  col = c('lightgrey', 'dodgerblue', "yellow1", 'brown4'), # Customize as needed
  row_title_gp = gpar(fontsize = 22), 
  column_title_gp = gpar(fontsize = 17), 
  row_split = row_split,  # Split rows by Lake
  column_split = column_split,           # Use the created column split
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_order = rownames(datamatrix), 
  #row_title_rot = 0,
  column_title_rot = 0,
  column_order = colnames(datamatrix),
  border = F,
  show_column_names = FALSE, 
  show_row_names = FALSE,
  show_row_dend = FALSE, 
  show_column_dend = FALSE,
  left_annotation = annot,
  #top_annotation = top_annotation,
  heatmap_legend_param = heatmap_legend_param
)
dev.off()

```
