# Nucleotide diversity (θ) plot across the whole genome and at putative inversion regions among Arctic charr morphs from Thingvallavatn homozygous for benthic and pelagic haplotype.

## Step 1: Load and Format Per-Window Thetas Data
```{r setup, include=FALSE}
#Libs
library(tidyverse)
library(patchwork)
library(zoo)
library(data.table)

#Directory
dir <- '~/Desktop/Comp_UU/REF_SalAlp_UK/Inversions/Nucleotide_diversity/ND_BvsP_maxDEPTH'

#Load data 
files <- list.files(dir, pattern = "pestPG$")


#Select desired files
#files_selected <- files[files %in% desired_files]


#Colnames
col_names <- c('Chr',	'WinCenter',	'tW',	'tP',	'tF',	'tH',	'tL',	'Tajima', 'fuf', 'fud',	'fayh',	'zeng',	'nSites')


# Read the desired files
thetas_list <- list()

for(file in files) {
  file_content <- read.table(file.path(dir, file), header = F)[,-1]
  colnames(file_content) <- col_names
  
  # Process the file content as needed
  thetas_list[[file]] <- file_content
}


#Modify names of the data list 
names(thetas_list) <- gsub("_pctind0.*_maxdepth.*500000\\.thetaswindow\\.pestPG$", '', names(thetas_list) )
```


## Step 2: Compute Summary Statistics Per Population
```{r}
# Assume `thetas_list` is your list
# Get the names of the list
list_names <- names(thetas_list)

# Define the target names
name_homB <- "ind_list_homB_chr_40_part_1.txt_inv_reg_chr40"
name_homP <- "ind_list_homP_chr_40_part_1.txt_inv_reg_chr40"

# Find indices of these names in the list
index_homB <- which(list_names == name_homB)
index_homP <- which(list_names == name_homP)

# Swap the names at these indices
if (length(index_homB) > 0 && length(index_homP) > 0) {
  list_names[index_homB] <- name_homP
  list_names[index_homP] <- name_homB
}

# Reassign the modified names to thetas_list
names(thetas_list) <- list_names


# Estimate Pi and Watterson’s theta  -------------------------------------------
l.theta <- list()
l.watt <- list()
l.sum <- list()
l.dfs <- list()

for (pop in seq_along(thetas_list)){
  df.theta <- thetas_list[[pop]]
  pop.name <- names(thetas_list[pop])
  l.theta[[pop.name]] <- mean((as.numeric(df.theta$tP) / as.numeric(df.theta$nSites)), na.rm = T)
  l.watt[[pop.name]] <- mean((as.numeric(df.theta$tW) /as.numeric(df.theta$nSites)), na.rm = T)
  l.sum[[pop.name]] <- sum(as.numeric(df.theta$nSites), na.rm = T)
  df.theta$pop <- pop.name
  l.dfs[[pop.name]] <- df.theta
}

length(l.theta) #as the number of populations
length(l.watt)  #as the number of populations

df.x2 <- data.frame(pop = names(l.theta),
                    pairwise.nuc = round(unlist(l.theta), 4),  
                    wattersons.theta = round(unlist(l.watt), 4),
                    number.sites = round(unlist(l.sum), 4))

```


## Step 3: Prepare Data for Visualization
```{r}
mean.theta.nsites <- df.x2
df.theta.nsites <- bind_rows(l.dfs)
names(df.theta.nsites) <- c("Chr", "WinCenter","tW", "tP","tF","tH","tL","Tajima","fuf","fud","fayh","zeng","nSites", "pop")


# cleaner summary ---------------------------------------------------------
#always filter windows > 10 sites
df.theta.nsites$nSites <- as.numeric(df.theta.nsites$nSites)
df.theta.nsites$tP <- as.numeric(df.theta.nsites$tP)
df.theta.nsites$tW <- as.numeric(df.theta.nsites$tW)

# Plot nucleotide diversity-----------------------------------------------------
df.theta.nsites_withPi <- df.theta.nsites
df.theta.nsites_withPi$Pi <-round(df.theta.nsites_withPi$tP/df.theta.nsites_withPi$nSites, 4)
df.theta.nsites_withPi <- df.theta.nsites_withPi[!is.na(df.theta.nsites_withPi$Pi),]


df.theta.nsites$WinCenter <- as.numeric(df.theta.nsites$WinCenter)
df.theta.nsites$WinCenter_less <- df.theta.nsites$WinCenter/1000000
df.theta.nsites$tP_per_site <- df.theta.nsites$tP/df.theta.nsites$nSites

#Clean data for the nSites , we are interested in nSites > 10
df.theta.nsites_clean <- df.theta.nsites[ df.theta.nsites$nSites > 10, ]

```



## Step 4: Annotate Inversion Regions and Morphs
```{r }

library(ggpubr)

#Combine all chromosomes and plot on one boxplot

df.theta.nsites_clean_boxplot <- 
  df.theta.nsites_clean %>% 
  dplyr::select(Chr, WinCenter_less, tP_per_site, pop)

df.theta.nsites_clean_boxplot <- 
  df.theta.nsites_clean_boxplot %>%
  mutate(
    Inversion = case_when(Chr == 1 & WinCenter_less >= 16.30 & WinCenter_less <= 18.60 ~ 'Inv1',
                               Chr == 1 & WinCenter_less >= 19.50 & WinCenter_less <= 22.20 ~ 'Inv1_2',
                               Chr == 3 & WinCenter_less >= 33.50 & WinCenter_less <= 35.80 ~ 'Inv3',
                               Chr == 3 & WinCenter_less >= 37.30 & WinCenter_less <= 40.60 ~ 'Inv3_2',
                               Chr == 8 & WinCenter_less >= 29.05 & WinCenter_less <= 29.83 ~ 'Inv8',
                               Chr == 9 & WinCenter_less >= 38.40 & WinCenter_less <= 40.80 ~ 'Inv9',
                               Chr == 14 & WinCenter_less >= 6.33 & WinCenter_less <= 6.87 ~ 'Inv14',
                               Chr == 40 & WinCenter_less >= 16.25 & WinCenter_less <= 17.01 ~ 'Inv40'
                               ),
        Group = case_when(
          
          startsWith(pop,'ind_list_homB') ~ 'Homozygous major (benthic)',
           startsWith(pop,'ind_list_homP') ~ 'Homozygous minor (pelagic)')
         )

df.theta.nsites_clean_boxplot_inv_only <- df.theta.nsites_clean_boxplot[!is.na(df.theta.nsites_clean_boxplot$Inversion),]

```


## Step 5: Perform Wilcoxon Tests per Inversion
```{r }
# List of all inversions to test
inversions <- c("Inv1", "Inv1_2", "Inv3", "Inv3_2", "Inv8", "Inv9", "Inv14", "Inv40")

data = df.theta.nsites_clean_boxplot_inv_only

# Initialize a data frame to store results
results <- data.frame(
  Inversion = character(),
  Morph1 = character(),
  Morph2 = character(),
  P_Value = numeric(),
  stringsAsFactors = FALSE
)

# Loop through each inversion
for (inv in inversions) {
  
  # Filter data for the two groups for the current inversion
  benthic <- data[data$Group == 'Homozygous major (benthic)' & data$Inversion == inv,]
  pelagic <- data[data$Group == 'Homozygous minor (pelagic)' & data$Inversion == inv,]
  
  # Perform Wilcoxon test
  test_result <- wilcox.test(benthic$tP_per_site, pelagic$tP_per_site)
  
  # Append the result to the results data frame
  results <- rbind(results, data.frame(
    Inversion = inv,
    Morph1 = "Homozygous major (benthic)",
    Morph2 = "Homozygous minor (pelagic)",
    P_Value = test_result$p.value
  ))
}

# Display results


```


## Step 6: Add summary statistics
```{r }
library(dplyr)

summary_inv <- 
  df.theta.nsites_clean_boxplot_inv_only %>%
  group_by(Inversion, pop) %>%
  summarise(mean_pi = round(mean(tP_per_site),4),
            sd_pi = round(sd(tP_per_site),4))

print(summary_inv)
```




## Step 7: Add data genome-wide
```{r }
#Directory
dir <- '~/Desktop/Comp_UU/REF_SalAlp_UK/Nucleotide_diversity/MORPH_PG'

#Load data 
files <- list.files(dir)

# Filter files
desired_files <- grep("maf_Thi_LB_SB|maf_Thi_Pi_PL", files, value = TRUE)
col_names <- c('Chr',	'WinCenter',	'tW',	'tP',	'tF',	'tH',	'tL',	'Tajima', 'fuf',
               'fud',	'fayh',	'zeng',	'nSites')

#I kb indeed are 50 Kbp, I just forgot to renames those
#Look at the list
#desired_files

# Read the desired files
thetas_list <- list()

for(file in desired_files) {
  file_content <- read.table(file.path(dir, file), header = F)[,-1]
  colnames(file_content) <- col_names
  # Process the file content as needed
  thetas_list[[file]] <- file_content
}

#Modify names of the data list 
names(thetas_list) <- gsub("_pctind0.*_maxdepth.*500000\\.thetaswindow\\.pestPG$", '', names(thetas_list) )
names(thetas_list) 

# List of prefixes to search for
prefixes <- c( 'maf_Thi_LB_SB', 'maf_Thi_Pi_PL')  # Add more prefixes as needed

# Initialize a new list to store combined data frames
new_data_list <- list()
prefix
prefixes
# Define a function to bind rows of data frames starting with a similar name
for(prefix in prefixes) {
  # Filter data frames starting with the specified prefix
  matching_data_frames <- grep(paste0("^", prefix), names(thetas_list), value = TRUE)
  
  # If there are matching data frames, bind them row-wise
  if (length(matching_data_frames) > 0) {
    combined_data <- do.call(rbind, thetas_list[matching_data_frames])
    new_data_list[[prefix]] <- combined_data
  }
}
combined_data
# Estimate Pi and Watterson’s theta  -------------------------------------------

l.theta <- list()
l.watt <- list()
l.sum <- list()
l.dfs <- list()

for (pop in seq_along(new_data_list)){
  df.theta <- new_data_list[[pop]]
  pop.name <- names(new_data_list[pop])
  l.theta[[pop.name]] <- mean((as.numeric(df.theta$tP) / as.numeric(df.theta$nSites)), na.rm = T)
  l.watt[[pop.name]] <- mean((as.numeric(df.theta$tW) /as.numeric(df.theta$nSites)), na.rm = T)
  l.sum[[pop.name]] <- sum(as.numeric(df.theta$nSites), na.rm = T)
  df.theta$pop <- pop.name
  l.dfs[[pop.name]] <- df.theta
}

length(l.theta) #as the number of populations
length(l.watt)  #as the number of populations

df.x2 <- data.frame(pop = names(l.theta),
                    pairwise.nuc = round(unlist(l.theta), 4), 
                    wattersons.theta = round(unlist(l.watt), 4),
                    number.sites = round(unlist(l.sum), 4))


mean.theta.nsites <- df.x2
df.theta.nsites <- bind_rows(l.dfs)
names(df.theta.nsites) <- c("Chr", "WinCenter","tW", "tP","tF","tH","tL","Tajima","fuf","fud","fayh","zeng","nSites", "pop")


# cleaner summary ---------------------------------------------------------
#always filter windows > 10 sites
df.theta.nsites$nSites <- as.numeric(df.theta.nsites$nSites)
df.theta.nsites$tP <- as.numeric(df.theta.nsites$tP)
df.theta.nsites$tW <- as.numeric(df.theta.nsites$tW)

clean.summary.thetas <- 
  df.theta.nsites %>% 
  filter(nSites > 10) %>%
  group_by(pop) %>%
  summarise(mean.pi = round(mean(tP / nSites),4),
            sd.pi = round(sd(tP / nSites),4),
            mean.wat = round(mean(tW / nSites),4),
            sd.wat = round(sd(tW / nSites),4)) %>%
  mutate(out.pi = paste0(mean.pi,"±",sd.pi),
         out.wat = paste0(mean.wat,"±",sd.wat),)

#write.csv(clean.summary.thetas, "/Users/khyku851/Desktop/REF_SalSp/Nucleotide_diversity/Output_thetas/clean_theta_folded_perpop.csv")

# Plot nucleotide diversity---------------------------------------

df.theta.nsites_withPi <- df.theta.nsites
df.theta.nsites_withPi$Pi <-round(df.theta.nsites_withPi$tP/df.theta.nsites_withPi$nSites, 4)

df.theta.nsites_withPi <- df.theta.nsites_withPi[!is.na(df.theta.nsites_withPi$Pi),]

#Use this data 
df.theta.nsites$WinCenter <- as.numeric(df.theta.nsites$WinCenter)
df.theta.nsites$WinCenter_less <- df.theta.nsites$WinCenter/1000000

df.theta.nsites$tP_per_site <- df.theta.nsites$tP/df.theta.nsites$nSites

#Clean data for the nSites , we are interested in nSites > 10
df.theta.nsites_clean_genome_wide <- df.theta.nsites[ df.theta.nsites$nSites > 10, ]

#Select and add group names
df.theta.nsites_clean_genome_wide <- df.theta.nsites_clean_genome_wide %>% 
  dplyr::select(Chr, WinCenter_less, tP_per_site, pop)

df.theta.nsites_clean_genome_wide$Inversion <- "Genome-wide"
 df.theta.nsites_clean_genome_wide <- 
   df.theta.nsites_clean_genome_wide %>% 
   mutate(Group = case_when(pop == 'maf_Thi_Pi_PL' ~ 'Pelagic',
                      pop == 'maf_Thi_LB_SB' ~ 'Benthic'))
 
 df.theta.nsites_clean_genome_wide$Chr <- 'Genome-wide'

 df.theta.nsites_clean_genome_wide_benthic <- df.theta.nsites_clean_genome_wide[df.theta.nsites_clean_genome_wide$Group == "Benthic",]
 df.theta.nsites_clean_genome_wide_pelagic <- df.theta.nsites_clean_genome_wide[df.theta.nsites_clean_genome_wide$Group == "Pelagic",]
 
wilcox.test(df.theta.nsites_clean_genome_wide_benthic$tP_per_site,df.theta.nsites_clean_genome_wide_pelagic$tP_per_site)
 
```


## Step 8: Plot boxplots
```{r}

df.theta.nsites_clean_genome_wide_homo <- rbind(df.theta.nsites_clean_genome_wide, df.theta.nsites_clean_boxplot_inv_only)

# Define your own colors for Group and Inversion
group_colors <- c("Benthic" =  "brown3", "Pelagic" = "darkgreen",
                  'Homozygous minor (pelagic)' =  "darkred", 
                  'Homozygous major (benthic)' = "#67a9cf")

            

inversion_colors <- c("Inv1" = "black", "Inv1_2" = "black", 
                      'Inv3' = "black", 'Inv3_2' = "black",
                      'Inv8' = "black", 'Inv9' = "black", 'Inv14' = "black", 
                      'Inv40' = "black", 
                      'Genome-wide' = "black")

# Assuming chr_levels is your previously defined levels for the x-axis (Chromosome)
chr_levels <- c("Genome-wide", "1", "3", "8", "9", "14", "40")

# Reorder the Chromosome variable in your dataframe using the previously defined levels
df.theta.nsites_clean_genome_wide_homo$Chr <- factor(df.theta.nsites_clean_genome_wide_homo$Chr, levels = chr_levels)


# Plot with custom colors
plt <- ggplot(df.theta.nsites_clean_genome_wide_homo) +
  geom_boxplot(aes(x = Chr, y = tP_per_site, color = Inversion, fill = Group), outlier.shape = NA) +
  scale_fill_manual(values = group_colors) +
  scale_color_manual(values = inversion_colors) +
  theme_classic(19) +
  theme(legend.position = 'none') +
  ylim(0, 0.007)+
  labs(x = 'Scaffold', y = "Theta")+
  annotate('text', y = 0.0035, x= 1, label="***", size = 5)+
  annotate('text', y = 0.0035, x= 1.8, label="*", size = 5)+
  annotate('text', y = 0.0028, x= 2.2, label="***", size = 5)+
  annotate('text', y = 0.0028, x= 2.8, label="***", size = 5)+
  annotate('text', y = 0.0028, x= 3.2, label="***", size = 5)+
  annotate('text', y = 0.002, x= 4, label="ns", size = 5)+
  annotate('text', y = 0.0018, x= 5, label="ns", size = 5)+
  annotate('text', y = 0.0018, x= 6, label="ns", size = 5)+
  annotate('text', y = 0.0051, x= 7, label="***", size = 5)+
  annotate("text", x = 1, y =  0.0058, label = "Benthic", color = "brown3", size = 5,
           fontface = "bold")+
  annotate("text", x = 1, y =  0.0048, label = "Pelagic", color = "darkgreen", size = 5,
           fontface = "bold")+
  annotate("text", x = 5, y =  0.0058, label = "Homozygous benthic", color = "#67a9cf", size = 5,fontface = "bold")+
  annotate("text", x = 5, y =  0.0048, label = "Homozygous pelagic", color = "brown4", size = 5, fontface = "bold")+
  #annotate("text", x = 5, y =  0.0067, label = "Wilcoxon test", color = "black", size = 5, fontface = "bold")+
    annotate("text", x = 0.5, y = 0.007, label = "h",size = 6, hjust = 0, vjust = 1, 
             fontface = "bold")


ggsave('~/Desktop/Comp_UU/REF_SalAlp_UK/Inversions/Output/Boxplots_nucl_div_HOMO_B_P.pdf', plt, width = 12, height= 3, dpi = 300)

```