## Step 1: Call genotypes to generate a set of SITES for the downstream analyses by using --minDepth  -setMaxDepth  filters 

```{r}
#!/bin/bash -l

#SBATCH -A uppmax2025-2-114
#SBATCH -p node
#SBATCH -n 1
#SBATCH -C mem256GB
#SBATCH --array=1-40
#SBATCH -t 08-00:00:00
#SBATCH -J GL_depth
#SBATCH -e GL_depth_%A_%a.err
#SBATCH -o GL_depth_%A_%a.out
#SBATCH --mail-type=all
#SBATCH --mail-user=khrystyna.kurta@slu.se


#Load modules
module load bioinfo-tools
module load ANGSD/0.933


#Path to the directory where you have the bam-files
BASEDIR=/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/Run_1_2
cd $BASEDIR

## Step 1: Define paths to Refference genome
REFGENOME=/proj/snic2020-2-19/private/arctic_charr/assemblies/fSalAlp1.1/Data_Package_fSalAlp1_assembly_20231024/fSalAlp1.1.hap1.cur.20231016.fasta
REF_INDEXED=/proj/snic2020-2-19/private/arctic_charr/assemblies/fSalAlp1.1/Data_Package_fSalAlp1_assembly_20231024/fSalAlp1.1.hap1.cur.20231016.fasta.fai

## Step 3: Create bam file list
BAMLIST=all_bam.list

### Step 2: Determine chromosome/ or Get all the contig (or scaffold) names from the reference genome fasta file
CHUNK_DIR=/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/CHUNK_LIST
CHUNK_LIST=/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/CHUNK_LIST/only_chr.list
CHUNK_ID=$(cat $CHUNK_LIST | sed -n ${SLURM_ARRAY_TASK_ID}p)
CHUNK_NAME=${CHUNK_ID/.txt/}


## Step 4: Specify PATH to sites 
SITES=/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/SITES/sites_sorted.txt

## Step 5: Specify OUTDIR
OUTDIR=/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/GENO_GL_DEPTH

#prepare variables - avoid to modify
NB_CPU=8
MIN_DEPTH=1

# advice: we usually set it at 2-4 times the expected coverage to remove repeated regions
MAX_DEPTH_FACTOR=3
MIN_MAF=0.05
PERCENT_IND=0.5
N_IND=$(cat $BAMLIST | wc -l)

# Calculate minimum number of individuals needed
MIN_IND_FLOAT=$(echo "($N_IND * $PERCENT_IND)" | bc -l)
MIN_IND=${MIN_IND_FLOAT%.*}  # Convert float to int (floor)
MAX_DEPTH=$(echo "($N_IND * $MAX_DEPTH_FACTOR)" |bc -l)
  
echo " output genotype in plink format for all individuals listed in $BAMLIST"
echo "work on SNPs listed in the sites file"


####Calculate the GL and generate ped plink files
angsd -P $NB_CPU -doMaf 1 -doCounts 1 \
    -GL 2 -doGeno 2 -doMajorMinor 4 -doPost 2 \
    -ref $REFGENOME -fai $REF_INDEXED \
    -rf $CHUNK_DIR/${CHUNK_ID}.txt \
    -minInd $MIN_IND -setMaxDepth $MAX_DEPTH -setMinDepthInd $MIN_DEPTH \
    -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -minMapQ 30 -minQ 20 \
    -sites $SITES \
    -bam $BAMLIST \
    -out $OUTDIR/Charr_RefUK_maf"$MIN_MAF"_pctind"$N_IND"_maxdepth"$MAX_DEPTH"_region"$CHUNK_NAME"


```

## Step 2: Merge beagle file, generate site list for the downstream analyses
```{r}
#!/bin/bash -l

#SBATCH -A naiss2024-5-277
#SBATCH -p core -n 10
#SBATCH -t 02-00:00:00
#SBATCH -J merge_b
#SBATCH -e merge_b_%A_%a.err
#SBATCH -o merge_b_%A_%a.out
#SBATCH --mail-type=all
#SBATCH --mail-user=khrystyna.kurta@slu.se

BEAGLE_DIR=/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/GENO_GL_DEPTH
SITE_DIR=/proj/snic2020-2-19/private/herring/users/khrystyna/Arctic_charr_ref_fSalAlp1/SITES/

cd $BEAGLE_DIR

# Step 1: Extract header from the first file
zcat $OUTDIR/${OUT_NAME}_maf"$MIN_MAF"_pctind"$N_IND"_maxdepth"$MAX_DEPTH"_region1 | head -n 1 > beagle_header.txt

# Step 2: Concatenate all files without headers into a temp file
for f in *.beagle.gz; do
    zcat "$f" | tail -n +2 >> beagle_body.txt
done

# Step 3: Combine header and body, then compress into final beagle.gz
cat beagle_header.txt beagle_body.txt | gzip > $OUTDIR/${OUT_NAME}_maf"$MIN_MAF"_pctind"$N_IND"_maxdepth"$MAX_DEPTH"_all_chr.beagle.gz

# Clean up intermediate files
rm beagle_header.txt beagle_body.txt

# Extract site positions
zcat *_all_chr.beagle.gz | cut -f1 | rev | sed 's/_/\t/' | rev > sites_depth.txt


mv sites_depth.txt $SITE_DIR

#Index sites
module load bioinfo-tools
module load ANGSD/0.933

cd $SITE_DIR

angsd sites index sites_depth.txt

```


